// @bryntum/calendar-react/lib/BryntumAgendaView.js
import React from 'react';
import { AgendaView } from '@bryntum/calendar';
import { createWidget, shouldComponentUpdate, processWidgetContent } from './WrapperHelper.js';
export class BryntumAgendaView extends React.Component {
    constructor() {
        super(...arguments);
        this.processWidgetContent = processWidgetContent;
    }
    componentDidMount() {
        this.instance = createWidget(this);
    }
    componentWillUnmount() {
        var _a, _b;
        (_b = (_a = this.instance) === null || _a === void 0 ? void 0 : _a.destroy) === null || _b === void 0 ? void 0 : _b.call(_a);
    }
    shouldComponentUpdate(nextProps, nextState) {
        return shouldComponentUpdate(this, nextProps, nextState);
    }
    render() {
        const className = `b-react-agenda-view-container`;
        return (React.createElement("div", { className: className, ref: (element) => (this.element = element) }));
    }
}
BryntumAgendaView.instanceClass = AgendaView;
BryntumAgendaView.instanceName = 'AgendaView';
BryntumAgendaView.configNames = [
    'adopt',
    'animateRemovingRows',
    'ariaDescription',
    'ariaLabel',
    'autoHeight',
    'bbar',
    'bodyCls',
    'bubbleEvents',
    'callOnFunctions',
    'collapsible',
    'config',
    'contentElementCls',
    'contextMenuTriggerEvent',
    'defaultRegion',
    'descriptionRenderer',
    'destroyStore',
    'disableGridColumnIdWarning',
    'disableGridRowModelWarning',
    'dock',
    'dragUnit',
    'enableSticky',
    'enableTextSelection',
    'endDate',
    'eventFilter',
    'eventHeaderRenderer',
    'eventHeight',
    'eventListTimeFormat',
    'eventRenderer',
    'eventRowSpacing',
    'eventSorter',
    'eventTimeRenderer',
    'fillLastColumn',
    'fixedRowHeight',
    'footer',
    'fullRowRefresh',
    'getPrimaryResource',
    'getRowHeight',
    'header',
    'hideEventOverflow',
    'hideHorizontalScrollbar',
    'hoverCls',
    'ignoreParentReadOnly',
    'listeners',
    'listRangeMenu',
    'loadMask',
    'loadMaskDefaults',
    'loadMaskError',
    'localizable',
    'maskDefaults',
    'masked',
    'monitorResize',
    'nonWorkingDays',
    'owner',
    'plugins',
    'preserveFocusOnDatasetChange',
    'preserveScrollOnDatasetChange',
    'preventTooltipOnTouch',
    'relayStoreEvents',
    'resizeToFitIncludesHeader',
    'resourceImagePath',
    'responsive',
    'responsiveLevels',
    'responsiveRoot',
    'responsiveState',
    'responsiveTarget',
    'ripple',
    'rootElement',
    'scrollerClass',
    'scrollManager',
    'shortEventCls',
    'shortEventDuration',
    'showBullet',
    'showDirty',
    'showResourceAvatars',
    'startDate',
    'stateful',
    'statefulEvents',
    'stateId',
    'stateProvider',
    'strips',
    'subGridConfigs',
    'syncMask',
    'tab',
    'tbar',
    'timeFormat',
    'type',
    'ui',
    'weekStartDay',
    'weight'
];
BryntumAgendaView.propertyConfigNames = [
    'alignSelf',
    'allowDragCreate',
    'allowDragMove',
    'allowDragResize',
    'alwaysShowCurrentDate',
    'animateTimeShift',
    'animateTreeNodeToggle',
    'appendTo',
    'ariaLive',
    'autoCreate',
    'cls',
    'collapsed',
    'columnLines',
    'columns',
    'data',
    'dataset',
    'date',
    'dateFilter',
    'dateFormat',
    'dateSeparator',
    'descriptionFormat',
    'disabled',
    'emptyText',
    'enableUndoRedoKeys',
    'extraData',
    'filterEventResources',
    'flex',
    'height',
    'hidden',
    'hideFooters',
    'hideHeaders',
    'hideNonWorkingDays',
    'id',
    'insertBefore',
    'insertFirst',
    'keyMap',
    'longPressTime',
    'margin',
    'maxDate',
    'maxHeight',
    'maxWidth',
    'minDate',
    'minHeight',
    'minWidth',
    'offsetStartsBeforeEvents',
    'onBeforeAutoCreate',
    'onBeforeCancelCellEdit',
    'onBeforeCellDelete',
    'onBeforeCellEditStart',
    'onBeforeChangeDate',
    'onBeforeColumnDragStart',
    'onBeforeColumnDropFinalize',
    'onBeforeColumnResize',
    'onBeforeCopy',
    'onBeforeCSVExport',
    'onBeforeDestroy',
    'onBeforeExcelExport',
    'onBeforeFillHandleDragStart',
    'onBeforeFinishCellEdit',
    'onBeforeHide',
    'onBeforePaste',
    'onBeforePdfExport',
    'onBeforeRenderRow',
    'onBeforeRenderRows',
    'onBeforeResponsiveStateChange',
    'onBeforeRowCollapse',
    'onBeforeRowExpand',
    'onBeforeSelectionChange',
    'onBeforeSetRecord',
    'onBeforeShow',
    'onBeforeStateApply',
    'onBeforeStateSave',
    'onBeforeToggleGroup',
    'onBeforeToggleNode',
    'onCancelCellEdit',
    'onCatchAll',
    'onCellClick',
    'onCellContextMenu',
    'onCellDblClick',
    'onCellMapPopulated',
    'onCellMenuBeforeShow',
    'onCellMenuItem',
    'onCellMenuShow',
    'onCellMenuToggleItem',
    'onCellMouseEnter',
    'onCellMouseLeave',
    'onCellMouseOut',
    'onCellMouseOver',
    'onCollapse',
    'onCollapseNode',
    'onColumnDrag',
    'onColumnDragStart',
    'onColumnDrop',
    'onColumnResize',
    'onColumnResizeStart',
    'onContextMenuItem',
    'onContextMenuToggleItem',
    'onCopy',
    'onDataChange',
    'onDayCellPopulated',
    'onDestroy',
    'onDirtyStateChange',
    'onDragSelecting',
    'onEventAutoCreated',
    'onEventPropagate',
    'onExpand',
    'onExpandNode',
    'onFileDrop',
    'onFillHandleBeforeDragFinalize',
    'onFillHandleDrag',
    'onFillHandleDragAbort',
    'onFillHandleDragEnd',
    'onFillHandleDragStart',
    'onFinishCellEdit',
    'onFocusIn',
    'onFocusOut',
    'onGridRowBeforeDragStart',
    'onGridRowBeforeDropFinalize',
    'onGridRowDrag',
    'onGridRowDragAbort',
    'onGridRowDragStart',
    'onGridRowDrop',
    'onHeaderClick',
    'onHeaderMenuBeforeShow',
    'onHeaderMenuItem',
    'onHeaderMenuShow',
    'onHeaderMenuToggleItem',
    'onHide',
    'onLockRows',
    'onMouseOut',
    'onMouseOver',
    'onPaint',
    'onPaste',
    'onPdfExport',
    'onRangeChange',
    'onReadOnly',
    'onRecompose',
    'onRefresh',
    'onRenderRow',
    'onRenderRows',
    'onResize',
    'onResponsive',
    'onResponsiveStateChange',
    'onRowCollapse',
    'onRowExpand',
    'onRowMouseEnter',
    'onRowMouseLeave',
    'onScroll',
    'onSelectionChange',
    'onSelectionModeChange',
    'onShow',
    'onSplit',
    'onSplitterCollapseClick',
    'onSplitterDragEnd',
    'onSplitterDragStart',
    'onSplitterExpandClick',
    'onStartCellEdit',
    'onSubGridCollapse',
    'onSubGridExpand',
    'onToggleGroup',
    'onToggleNode',
    'onToolClick',
    'onTopDateChange',
    'onUnlockRows',
    'onUnsplit',
    'preserveScroll',
    'range',
    'readOnly',
    'rowHeight',
    'rowLines',
    'scrollable',
    'selectionMode',
    'shiftIncrement',
    'shortDateFormat',
    'shortDateTimeFormat',
    'store',
    'syncCalendarDate',
    'syncViewDate',
    'title',
    'tools',
    'transitionDuration',
    'width'
];
BryntumAgendaView.propertyNames = [
    'accessibility',
    'dayCellCls',
    'firstVisibleCell',
    'firstVisibleDate',
    'hasChanges',
    'lastVisibleCell',
    'lastVisibleDate',
    'originalStore',
    'parent',
    'selectedCell',
    'selectedCells',
    'selectedRecord',
    'selectedRecords',
    'selectedRows',
    'state',
    'tooltip'
];
