// @bryntum/calendar-react/src/BryntumCalendar.tsx
/**
 * React wrapper for Bryntum Calendar
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, { RefObject } from 'react';
import { AgendaViewConfig, AlignSpec, AssignmentModel, AssignmentModelConfig, AssignmentStore, AssignmentStoreConfig, Base, ButtonConfig, Calendar, CalendarDatePickerConfig, CalendarDrag, CalendarDragConfig, CalendarListeners, CalendarMixin, Column, Container, ContainerConfig, ContainerItemConfig, ContainerLayoutConfig, CrudManagerConfig, DatePickerConfig, DayCell, DayResourceViewConfig, DayViewConfig, DragContext, EventBuffer, EventBufferConfig, EventCopyPaste, EventCopyPasteConfig, EventEdit, EventEditConfig, EventListConfig, EventMenu, EventMenuConfig, EventModel, EventModelConfig, EventStore, EventStoreConfig, EventTooltip, EventTooltipConfig, ExcelExporter, ExcelExporterConfig, ExportConfig, ExternalEventSource, ExternalEventSourceConfig, KeyMapConfig, Layout, LoadOnDemand, LoadOnDemandConfig, Mask, MaskConfig, Menu, MenuItem, MenuItemConfig, Model, MonthViewConfig, PagingToolbarConfig, Panel, PanelCollapserConfig, PanelCollapserOverlayConfig, PanelHeader, Popup, Print, PrintConfig, ProjectConsumer, ProjectModel, ProjectModelConfig, ProjectModelMixin, Rectangle, RecurrenceConfirmationPopup, RecurrenceConfirmationPopupConfig, ResourceFilterConfig, ResourceModel, ResourceModelConfig, ResourceStore, ResourceStoreConfig, ResourceTimeRangeModel, ResourceTimeRangeModelConfig, ResourceTimeRangeStore, ResourceTimeRangeStoreConfig, ResourceViewConfig, ScheduleMenu, ScheduleMenuConfig, ScheduleTooltip, ScheduleTooltipConfig, Scheduler, SchedulerCrudManager, SchedulerEventEdit, Scroller, ScrollerConfig, StateProvider, Store, TabConfig, TextFieldConfig, TimeRangeModel, TimeRangeModelConfig, TimeRangeStore, TimeRangeStoreConfig, TimeRanges, TimeRangesConfig, TimeSpan, Tool, ToolConfig, ToolbarConfig, TooltipConfig, ValidateCreateResult, WeekExpander, WeekExpanderConfig, WeekViewConfig, Widget, XLSColumn, YearViewConfig } from '@bryntum/calendar';

import { createWidget, shouldComponentUpdate, processWidgetContent } from './WrapperHelper.js';

export type BryntumCalendarProps = {
    // Configs
    /**
     * An optional CSS class name to add to calendar date cells which encapsulate the calendar's
     * active [date](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#property-date).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-activeDateCls)
     */
    activeDateCls? : string
    /**
     * Element (or element id) to adopt as this Widget's encapsulating element. The widget's
     * content will be placed inside this element.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-adopt)
     */
    adopt? : HTMLElement|string
    /**
     * *Only valid if this Widget is [floating](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-floating).*
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-align)
     */
    align? : AlignSpec|string
    /**
     * When this widget is a child of a [Container](https://bryntum.com/products/calendar/docs/api/Core/widget/Container), it will by default be participating in a
     * flexbox layout. This config allows you to set this widget's
     * [align-self](https://developer.mozilla.org/en-US/docs/Web/CSS/align-self) style.
     */
    alignSelf? : string
    /**
     * *Only valid if this Widget is [floating](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-floating) and being shown through [showBy](#Core/widget/Widget#function-showBy).*
     * `true` to show a connector arrow pointing to the align target.
     */
    anchor? : boolean
    /**
     * Element (or the id of an element) to append this widget's element to. Can be configured, or set once at
     * runtime. To access the element of a rendered widget, see [element](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-element).
     */
    appendTo? : HTMLElement|string
    /**
     * A localizable string (May contain `'L{}'` tokens which resolve in the locale file) to inject
     * into an element which will be linked using the `aria-describedby` attribute.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-ariaDescription)
     */
    ariaDescription? : string
    /**
     * A localizable string (May contain `'L{}'` tokens which resolve in the locale file) to inject as
     * the `aria-label` attribute.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-ariaLabel)
     */
    ariaLabel? : string
    /**
     * The `aria-live` value for this widget.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-ariaLive)
     */
    ariaLive? : string
    /**
     * Inline assignments, will be loaded into the [assignmentStore](https://bryntum.com/products/calendar/docs/api/Calendar/mixin/CalendarStores#property-assignmentStore).
     */
    assignments? : AssignmentModel[]|AssignmentModelConfig[]
    /**
     * The [AssignmentStore](https://bryntum.com/products/calendar/docs/api/Scheduler/data/AssignmentStore) or a reconfiguring object for a store which will hold assignments linking resources to events.
     */
    assignmentStore? : AssignmentStore|AssignmentStoreConfig
    /**
     * If this config is set, then the `gesture` configured (which defaults to `dblclick`) creates a
     * new event at the mouse or touch event's time point.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-autoCreate)
     */
    autoCreate? : {
        gesture?: string
        newName?: Function|string
        step?: string
        duration?: string
        startHour?: number|string
    }|string|boolean
    /**
     * A Config object representing the configuration of a [Toolbar](https://bryntum.com/products/calendar/docs/api/Core/widget/Toolbar),
     * or array of config objects representing the child items of a Toolbar. Another way to add a bbar is to use [strips](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-strips).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-bbar)
     */
    bbar? : (ContainerItemConfig|string)[]|ToolbarConfig|PagingToolbarConfig|null
    /**
     * Custom CSS classes to add to the panel's body element.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-bodyCls)
     */
    bodyCls? : string|object
    /**
     * An object where property names with a truthy value indicate which events should bubble up the ownership
     * hierarchy when triggered.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-bubbleEvents)
     */
    bubbleEvents? : object
    /**
     * Set to `false` to not call onXXX method names (e.g. `onShow`, `onClick`), as an easy way to listen for events.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-callOnFunctions)
     */
    callOnFunctions? : boolean
    /**
     * *Only valid if this Widget is [floating](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-floating).*
     * Set to `true` to centre the Widget in browser viewport space.
     */
    centered? : boolean
    /**
     * Custom CSS classes to add to element.
     * May be specified as a space separated string, or as an object in which property names
     * with truthy values are used as the class names:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-cls)
     */
    cls? : string|object
    /**
     * Controls whether the panel is collapsed (the body of the panel is hidden while only the header is
     * visible). Only valid if the panel is [collapsible](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-collapsible).
     */
    collapsed? : boolean
    /**
     * This config enables collapsibility for the panel. See [collapsed](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-collapsed).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-collapsible)
     */
    collapsible? : boolean|PanelCollapserConfig|PanelCollapserOverlayConfig
    config? : object
    /**
     * *Only valid if this Widget is [floating](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-floating) or [positioned](#Core/widget/Widget#config-positioned).*
     * Element, Widget or Rectangle to which this Widget is constrained.
     */
    constrainTo? : HTMLElement|Widget|Rectangle
    /**
     * Custom CSS classes to add to the [contentElement](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-contentElement).
     * May be specified as a space separated string, or as an object in which property names
     * with truthy values are used as the class names:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-contentElementCls)
     */
    contentElementCls? : string|object
    /**
     * Event which is used to show context menus via the context menu features
     * [EventMenu](https://bryntum.com/products/calendar/docs/api/Calendar/feature/EventMenu) and [ScheduleMenu](#Calendar/feature/ScheduleMenu).
     * Available options are: 'contextmenu', 'click', 'dblclick'.
     */
    contextMenuTriggerEvent? : 'contextmenu'|'click'|'dblclick'
    /**
     * Supply a [CrudManager](https://bryntum.com/products/calendar/docs/api/Scheduler/data/CrudManager) instance or a config object if you want to use
     * CrudManager for handling data.
     */
    crudManager? : CrudManagerConfig|SchedulerCrudManager
    /**
     * Class that should be used to instantiate a CrudManager in case it's provided as a simple object to
     * [crudManager](https://bryntum.com/products/calendar/docs/api/Calendar/mixin/CalendarStores#config-crudManager) config.
     */
    crudManagerClass? : SchedulerCrudManager
    /**
     * Object to apply to elements dataset (each key will be used as a data-attribute on the element)
     */
    dataset? : Record<string, string>
    /**
     * The date which the Calendar, its [sidebar](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#property-sidebar)'s date picker, and
     * its active view should encapsulate upon creation.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-date)
     */
    date? : Date|string
    /**
     * A [DateHelper](https://bryntum.com/products/calendar/docs/api/Core/helper/DateHelper) format string to use to create date output for
     * view descriptions.
     */
    dateFormat? : string
    /**
     * An optional config object to configure the [CalendarDatePicker](https://bryntum.com/products/calendar/docs/api/Calendar/widget/CalendarDatePicker) which is
     * shown in the [sidebar](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#property-sidebar) next to the calendar views.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-datePicker)
     * @param {boolean|'count'|'dots'} datePicker.showEvents Displays the presence of events in each cell. values may be:  * `false` - Do not show events in cells. * `true` - Show a themeable bullet to indicate the presence of events for a date. * `'count'` - Show a themeable badge containing the event count for a date. * '`dots'` - Show small event-coloured bullets (to a maximum of three) below the date.
     * @param {boolean} datePicker.collapsible `true` to enable collapse into a header.
     * @param {string} datePicker.title The text to place in a header above the picker.
     * @param {boolean} datePicker.editMonth `true` to enable setting the date using the controls in the DatePicker header.
     * @param {boolean|'range'} multiSelect The DatePicker can select [multiple dates](https://bryntum.com/products/calendar/docs/api/Core/widget/DatePicker#config-multiSelect) in two modes:  * `true` to mean many discontiguous dates may be selected. * `'range'` to mean that a start date and end date may be selected.  If set up to multi select, views which are not fixed to a date duration will obey the date selection by using a [dateFilter](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView#property-dateFilter) to skip days which are not selected. These include:  * AgendaView      (mode type `'agenda'`) * DayViews        (view type `'dayview'`) which you add to the modes - The default `day` and `week` modes have a fixed duration. * ResourceView    (view type `'resourceview'`) * DayResourceView (view type `'dayresourceview'`)  The `MonthView`, and the `WeekView` and the built-in single day `DayView` <strong>mode</strong> all have fixed duration and will not participate in this. If you select a discontiguous range, then the views which are amenable to being controlled in this way will switch their range to encompass what you have selected. For example, if you select the first to the last days of a month, the view's [range](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DateRangeOwner#property-range) will be set to `'1 month'`. This choice will be maintained, and by navigating using the next and previous tbar buttons, the views will move by exactly one month. If your set of selected dates is enclosed within a two week block, then the [range](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DateRangeOwner#property-range) of the views will be set to `{ unit : 'week', magnitude : 2 }`. If you select the 1st, 3th and 5th of the month, and those days are all on one week row, the [range](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DateRangeOwner#property-range) of the views will be set to `{ unit : 'week', magnitude : 1 }`. For selection ranges other than exactly one month, navigation in time will be in increments of one week for intuitive navigation up and down the date picker. When navigating in time, the selected block of dates moves and *follows the time navigation* so that there are always dates in view, even if they are empty.
     */
    datePicker? : CalendarDatePickerConfig|boolean|null
    /**
     * The name of the property to set when a single value is to be applied to this Widget. Such as when used
     * in a grid WidgetColumn, this is the property to which the column's `field` is applied.
     */
    defaultBindProperty? : string
    /**
     * The calendar ([Resource](https://bryntum.com/products/calendar/docs/api/Scheduler/model/ResourceModel)), (or its `id`) to use as the
     * default calendar for new events created using [dblclick](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-autoCreate)
     * or [drag-create](https://bryntum.com/products/calendar/docs/api/Calendar/feature/CalendarDrag), or [dragging in](#Calendar/feature/ExternalEventSource).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-defaultCalendar)
     */
    defaultCalendar? : string|ResourceModel
    /**
     * A [query](https://bryntum.com/products/calendar/docs/api/Core/widget/Container#function-query) selector function which can identify the descendant widget to which
     * focus should be directed by default.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-defaultFocus)
     * @param {Widget} widget Widget passed to method
     * @returns {boolean} truthy value if widget is the default one
     */
    defaultFocus? : (widget: Widget) => boolean
    /**
     * A config object containing default settings to apply to all child widgets.
     */
    defaults? : ContainerItemConfig
    /**
     * Configure as `false` to preserve selection when clicking the empty schedule area.
     */
    deselectAllOnScheduleClick? : boolean
    /**
     * Configure as `true` to deselect a selected event upon click.
     */
    deselectOnClick? : boolean
    /**
     * Configure as `true` to destroy the Project and stores when `this` is destroyed.
     */
    destroyStores? : boolean
    /**
     * Disable or enable the widget. It is similar to [readOnly](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-readOnly) except a disabled widget
     * cannot be focused, uses a different rendition (usually greyish) and does not allow selecting its value.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-disabled)
     */
    disabled? : boolean|'inert'
    /**
     * Controls the placement of this widget when it is added to a [panel's ](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel)
     * [strips collection](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-strips). Typical values for this config are `'top'`,
     * `'bottom'`, `'left'`, or `'right'`, which cause the widget to be placed on that side of the panel's
     * body. Such widgets are called "edge strips".
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-dock)
     */
    dock? : 'top'|'bottom'|'left'|'right'|'start'|'end'|'header'|'pre-header'|object
    /**
     * *Only valid if this Widget is [floating](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-floating).*
     * Set to `true` to be able to drag a widget freely on the page. Or set to an object with a ´handleSelector´
     * property which controls when a drag should start.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-draggable)
     */
    draggable? : boolean|{
        handleSelector?: string
    }
    /**
     * Allow using [Delete] and [Backspace] to remove events/assignments
     */
    enableDeleteKey? : boolean
    /**
     * Enables events being recurring and also adds extra recurrence UI fields in the built-in event editor.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-enableRecurringEvents)
     */
    enableRecurringEvents? : boolean|{
        defaultAction?: 'single'|'future'
    }
    /**
     * Set to `true` to listen for CTRL-Z (CMD-Z on Mac OS) keyboard event and trigger undo (redo when SHIFT is pressed).
     * Only applicable when using a [StateTrackingManager](https://bryntum.com/products/calendar/docs/api/Core/data/stm/StateTrackingManager).
     */
    enableUndoRedoKeys? : boolean
    /**
     * Inline events, will be loaded into the [eventStore](https://bryntum.com/products/calendar/docs/api/Calendar/mixin/CalendarStores#property-eventStore).
     */
    events? : EventModel[]|EventModelConfig[]
    /**
     * Configure as `true`, or set property to `true` to disable event selection.
     */
    eventSelectionDisabled? : boolean
    /**
     * The [EventStore](https://bryntum.com/products/calendar/docs/api/Scheduler/data/EventStore) or a reconfiguring object for a store which will hold
     * the events to be rendered into the Calendar.
     */
    eventStore? : EventStore|EventStoreConfig
    extraData? : any
    /**
     * If this is set to `true`, then when determining which assigned resource of a multi assigned event
     * to use to create the event UI, the first resource which is still selected in the
     * [resourceFilter](https://bryntum.com/products/calendar/docs/api/Calendar/widget/Sidebar#property-resourceFilter) is used.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-filterEventResources)
     */
    filterEventResources? : boolean
    /**
     * When this widget is a child of a [Container](https://bryntum.com/products/calendar/docs/api/Core/widget/Container), it will by default be participating in a
     * flexbox layout. This config allows you to set this widget's
     * [flex](https://developer.mozilla.org/en-US/docs/Web/CSS/flex) style.
     * This may be configured as a single number or a `&lt;flex-grow&gt; &lt;flex-shrink&gt; &lt;flex-basis&gt;` format string.
     * numeric-only values are interpreted as the `flex-grow` value.
     */
    flex? : number|string
    /**
     * Set to `true` to move the widget out of the document flow and position it
     * absolutely in browser viewport space.
     */
    floating? : boolean
    /**
     * Config object of a footer. May contain a `dock`, `html` and a `cls` property. A footer is not a widget,
     * but rather plain HTML that follows the last element of the panel's body and [strips](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-strips).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-footer)
     */
    footer? : {
        dock?: 'top'|'right'|'bottom'|'left'|'start'|'end'
        html?: string
        cls?: string
    }|string
    /**
     * A config [object](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#typedef-PanelHeader) for the panel's header or a string in place of a `title`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-header)
     */
    header? : string|boolean|PanelHeader
    /**
     * Widget's height, used to set element style.height. Either specify a valid height string or a number, which
     * will get 'px' appended. We recommend using CSS as the primary way to control height, but in some cases
     * this config is convenient.
     */
    height? : string|number
    /**
     * Configure with true to make widget initially hidden.
     */
    hidden? : boolean
    /**
     * *Only valid if this Widget is [floating](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-floating).*
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-hideAnimation)
     */
    hideAnimation? : boolean|object
    /**
     * Configure as `true` to hide [nonWorkingDays](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-nonWorkingDays) for each calendar view
     */
    hideNonWorkingDays? : boolean
    /**
     * Specify `true` to make this container hide when it has no visible children (Either empty
     * or all children hidden).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-hideWhenEmpty)
     */
    hideWhenEmpty? : boolean
    /**
     * Configure this as `true` to make picking a date in the [sidebar](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-sidebar)'s
     * date picker highlight the date cell in the active view.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-highlightDate)
     * @param {Date} date Cell date
     * @returns {boolean} Returns `true` to make picking a date in a date picker
     */
    highlightDate? : boolean|((date: Date) => boolean)
    /**
     * Configure as `true`, or set property to `true` to highlight dependent events as well when selecting an event.
     */
    highlightPredecessors? : boolean
    /**
     * Configure as `true`, or set property to `true` to highlight dependent events as well when selecting an event.
     */
    highlightSuccessors? : boolean
    /**
     * Widget id, if not specified one will be generated. Also used for lookups through Widget.getById
     */
    id? : string
    /**
     * Determines if the widgets read-only state should be controlled by its parent.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-ignoreParentReadOnly)
     */
    ignoreParentReadOnly? : boolean
    /**
     * Set to `true` to show the "Weekends" toggle button of the mode selector. Alternatively, this config
     * can be a [Button](https://bryntum.com/products/calendar/docs/api/Core/widget/Button) config object to configure the button.
     */
    includeWeekendsButton? : boolean|ButtonConfig
    /**
     * Element (or element id) to insert this widget before. If provided, [appendTo](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-appendTo) config is ignored.
     */
    insertBefore? : HTMLElement|string
    /**
     * Element (or element id) to append this widget element to, as a first child. If provided, [appendTo](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-appendTo) config is ignored.
     */
    insertFirst? : HTMLElement|string
    /**
     * A template method (empty by default) allowing you to control if an event can be selected or not.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-isEventSelectable)
     * @param {EventModel} event The event record
     * @returns {boolean} true if event can be selected, otherwise false
     */
    isEventSelectable? : (event: EventModel) => boolean
    /**
     * An optional CSS class to add to child items of this container.
     */
    itemCls? : string
    /**
     * An object containing typed child widget config objects or Widgets. May also be specified
     * as an array.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-items)
     */
    items? : Record<string, ContainerItemConfig|MenuItemConfig|boolean|null>|(ContainerItemConfig|MenuItemConfig|Widget)[]
    /**
     * An object whose keys are the [key](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) name
     * and optional modifier prefixes: `'Ctrl+'`, `'Alt+'`, `'Meta+'`, and `'Shift+'` (case-insensitive). The values
     * are the name of the instance method to call when the keystroke is received.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-keyMap)
     */
    keyMap? : Record<string, KeyMapConfig>
    /**
     * The short name of a helper class which manages rendering and styling of child items.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-layout)
     */
    layout? : string|ContainerLayoutConfig
    /**
     * The CSS style properties to apply to the [contentElement](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-contentElement).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-layoutStyle)
     */
    layoutStyle? : object
    /**
     * An array of [child item](https://bryntum.com/products/calendar/docs/api/Core/widget/Container#config-items) *config objects* which is to be converted into
     * instances only when this Container is rendered, rather than eagerly at construct time.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-lazyItems)
     */
    lazyItems? : Record<string, ContainerItemConfig>|ContainerItemConfig[]|Widget[]
    /**
     * The listener set for this object.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-listeners)
     */
    listeners? : CalendarListeners
    /**
     * A [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object, or a message to be shown when a store is performing a remote
     * operation, or Crud Manager is loading data from the sever. Set to `null` to disable default load mask.
     */
    loadMask? : string|MaskConfig|null
    /**
     * A [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object to adjust the [maskDefaults](#Core/widget/Widget#config-maskDefaults)
     * when data is loading. The message and optional configuration from the
     * [loadMask](https://bryntum.com/products/calendar/docs/api/Core/mixin/LoadMaskable#config-loadMask) config take priority over these options, just as they do
     * for `maskDefaults`, respectively.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-loadMaskDefaults)
     */
    loadMaskDefaults? : MaskConfig
    /**
     * A [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object to adjust the [maskDefaults](#Core/widget/Widget#config-maskDefaults)
     * when an error occurs loading data.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-loadMaskError)
     */
    loadMaskError? : MaskConfig|Mask|boolean
    /**
     * A class translations of which are used for translating this entity.
     * This is often used when translations of an item are defined on its container class.
     * For example:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-localeClass)
     */
    localeClass? : typeof Base
    /**
     * Set to `false` to disable localization of this object.
     */
    localizable? : boolean
    /**
     * List of properties which values should be translated automatically upon a locale applying.
     * In case there is a need to localize not typical value (not a String value or a field with re-defined setter/getter),
     * you could use 'localeKey' meta configuration.
     * Example:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-localizableProperties)
     */
    localizableProperties? : string[]
    /**
     * This flag controls whether Scheduler should preserve its selection of events when loading a new dataset
     * (if selected event ids are included in the newly loaded dataset).
     */
    maintainSelectionOnDatasetChange? : boolean
    /**
     * Widget's margin. This may be configured as a single number or a `TRBL` format string.
     * numeric-only values are interpreted as pixels.
     */
    margin? : number|string
    /**
     * This config object contains the defaults for the [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) created for the
     * [masked](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-masked) config. Any properties specified in the `masked` config will override these
     * values.
     */
    maskDefaults? : MaskConfig
    /**
     * Set to `true` to apply the default mask to the widget. Alternatively, this can be the mask message or a
     * [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object.
     */
    masked? : boolean|string|MaskConfig
    /**
     * The maximum date to which the `endDate` of any child view may be navigated.
     */
    maxDate? : Date|string
    /**
     * The element's maxHeight. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [height](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-height), *reading* the value will return the numeric value in pixels.
     */
    maxHeight? : string|number
    /**
     * *Only valid if this Widget is [floating](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-floating).*
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-maximizeOnMobile)
     */
    maximizeOnMobile? : number|string
    /**
     * The elements maxWidth. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [width](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-width), *reading* the value will return the numeric value in pixels.
     */
    maxWidth? : string|number
    /**
     * The minimum date to which the `startDate` of any child view may be navigated.
     */
    minDate? : Date|string
    /**
     * The element's minHeight. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [height](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-height), *reading* the value will return the numeric value in pixels.
     */
    minHeight? : string|number
    /**
     * The elements minWidth. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [width](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-width), *reading* the value will return the numeric value in pixels.
     */
    minWidth? : string|number
    /**
     * Which of the built-in [modes](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-modes) should be the initially active view.
     */
    mode? : 'agenda'|'year'|'month'|'week'|'day'|'list'|'resource'|'dayresource'|string
    /**
     * An object to use to configure common properties for all [modes](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-modes) prior to their construction.
     */
    modeDefaults? : object
    /**
     * Configures the view modes which are to be available in this Calendar.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-modes)
     */
    modes? : {
        [mode: string]: ContainerItemConfig|boolean|null|undefined
        agenda?: AgendaViewConfig|boolean|null
        year?: YearViewConfig|boolean|null
        month?: MonthViewConfig|boolean|null
        week?: WeekViewConfig|boolean|null
        day?: DayViewConfig|boolean|null
        list?: EventListConfig|boolean|null
        resource?: ResourceViewConfig|boolean|null
        dayresourceview?: DayResourceViewConfig|boolean|null
    }
    /**
     * When this is configured as `true` a [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver)
     * is used to monitor this element for size changes caused by either style manipulation, or by CSS
     * layout.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-monitorResize)
     */
    monitorResize? : boolean
    /**
     * Configure as `true` to allow `CTRL+click` to select multiple events in the scheduler.
     */
    multiEventSelect? : boolean
    /**
     * An object containing default config objects which may be referenced by name in the [items](https://bryntum.com/products/calendar/docs/api/Core/widget/Container#config-items)
     * config. For example, a specialized [Menu](https://bryntum.com/products/calendar/docs/api/Core/widget/Menu) subclass may have a `namedItems` default
     * value defined like this:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-namedItems)
     */
    namedItems? : Record<string, ContainerItemConfig>
    /**
     * Indicates where the Next/Previous/Today buttons should be placed.
     */
    navigatorPlacement? : 'toolbar'|'sidebar'
    /**
     * Non-working days as an object where keys are day indices, 0-6 (Sunday-Saturday), and the value is `true`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-nonWorkingDays)
     */
    nonWorkingDays? : Record<number, boolean>
    /**
     * By default, the [sidebar](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#property-sidebar) expands and collapses taking width from the
     * calendar UI. Configure `overlaySidebar` as `true` to have it start collapsed, and then
     * <strong>overlay</strong> the calendar UI when the expand/collapse button is toggled.
     */
    overlaySidebar? : boolean
    /**
     * The owning Widget of this Widget. If this Widget is directly contained (that is, it is one of the
     * [items](https://bryntum.com/products/calendar/docs/api/Core/widget/Container#property-items) of a Container), this config will be ignored. In this case
     * the owner is <strong>always</strong> the encapsulating Container.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-owner)
     */
    owner? : Widget
    /**
     * Set to `true` when a widget is rendered into another widget's [contentElement](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-contentElement), but must
     * not participate in the standard layout of that widget, and must be positioned relatively to that
     * widget's [contentElement](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-contentElement).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-positioned)
     */
    positioned? : boolean
    /**
     * Prevent tooltip from being displayed on touch devices. Useful for example for buttons that display a
     * menu on click etc, since the tooltip would be displayed at the same time.
     */
    preventTooltipOnTouch? : boolean
    /**
     * A [ProjectModel](https://bryntum.com/products/calendar/docs/api/Scheduler/model/ProjectModel) instance or a config object. The project holds all Scheduler data.
     * Can be omitted in favor of individual store configs or [crudManager](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/SchedulerStores#config-crudManager) config.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-project)
     */
    project? : ProjectModel|ProjectModelConfig|RefObject<any>
    /**
     * Configure as `true` to make the calendar read-only, by disabling any UIs for modifying data.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-readOnly)
     */
    readOnly? : boolean
    /**
     * The confirmation dialog shown when a recurring event is edited.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-recurrenceConfirmationPopup)
     */
    recurrenceConfirmationPopup? : RecurrenceConfirmationPopupConfig|RecurrenceConfirmationPopup
    relayStoreEvents? : boolean
    /**
     * Path to load resource images from. Used by the resource headers in ResourceView and by
     * [showResourceAvatars](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/CalendarMixin#config-showResourceAvatars) in event-displaying
     * modes.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-resourceImagePath)
     */
    resourceImagePath? : string
    /**
     * Inline resources, will be loaded into the [resourceStore](https://bryntum.com/products/calendar/docs/api/Calendar/mixin/CalendarStores#property-resourceStore).
     */
    resources? : ResourceModel[]|ResourceModelConfig[]
    /**
     * The [ResourceStore](https://bryntum.com/products/calendar/docs/api/Scheduler/data/ResourceStore) or a reconfiguring object for a store which will hold the resources to be rendered into the Calendar.
     */
    resourceStore? : ResourceStore|ResourceStoreConfig
    /**
     * Inline resource time ranges, will be loaded into [resourceTimeRangeStore](https://bryntum.com/products/calendar/docs/api/Calendar/mixin/CalendarStores#property-resourceTimeRangeStore).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-resourceTimeRanges)
     */
    resourceTimeRanges? : ResourceTimeRangeModel[]|ResourceTimeRangeModelConfig[]
    /**
     * The [ResourceTimeRangeStore](https://bryntum.com/products/calendar/docs/api/Scheduler/data/ResourceTimeRangeStore) or a reconfiguring object for a store which will hold resource time ranges.
     * These may be used by the [TimeRanges](https://bryntum.com/products/calendar/docs/api/Calendar/feature/TimeRanges) feature.
     */
    resourceTimeRangeStore? : ResourceTimeRangeStore|ResourceTimeRangeStoreConfig
    /**
     * Specifies the various responsive state objects keyed by their name. Each key (except `'*'`, see below) in
     * this object is a state name (see [responsiveState](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Responsive#config-responsiveState)) and its corresponding value is the
     * associated [ResponsiveState](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Responsive#typedef-ResponsiveState) object.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-responsive)
     */
    responsive? : object
    /**
     * Set to `true` to mark this instance as the default [responsiveTarget](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Responsive#config-responsiveTarget) for descendants that do
     * not specify an explicit [responsiveTarget](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Responsive#config-responsiveTarget) of their own.
     */
    responsiveRoot? : boolean
    /**
     * The name of the active state of the [responsive](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Responsive#config-responsive) config. This is assigned internally
     * and should not be assigned directly.
     */
    responsiveState? : string
    /**
     * The widget whose size and other properties drive this object's responsive behavior. If this config is not
     * specified, the closest ancestor that specified [responsiveRoot=true](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Responsive#config-responsiveRoot) will be
     * used. If there is no such ancestor, then the instance using this mixin is used.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-responsiveTarget)
     */
    responsiveTarget? : string|Widget
    /**
     * Configure as `true` to have the component display a translucent ripple when its
     * [focusElement](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-focusElement), or [element](#Core/widget/Widget#property-element) is tapped <em>if the
     * current theme supports ripples</em>. Out of the box, only the Material theme supports ripples.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-ripple)
     */
    ripple? : boolean|{
        delegate?: string
        color?: string
        radius?: number
        clip?: string
    }
    /**
     * If you are rendering this widget to a shadow root inside a web component, set this config to the shadowRoot. If not inside a web component, set it to `document.body`
     */
    rootElement? : ShadowRoot|HTMLElement
    /**
     * Specifies whether (and optionally in which axes) a Widget may scroll. `true` means this widget may scroll
     * in both axes. May be an object containing boolean `overflowX` and `overflowY` properties which are
     * applied to CSS style properties `overflowX` and `overflowY`. If they are boolean, they are translated to
     * CSS overflow properties thus:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-scrollable)
     */
    scrollable? : boolean|ScrollerConfig|Scroller
    /**
     * Defines what to do if document is scrolled while Widget is visible (only relevant when floating is set to `true`).
     * Valid values: ´null´: do nothing, ´hide´: hide the widget or ´realign´: realign to the target if possible.
     */
    scrollAction? : 'hide'|'realign'|null
    /**
     * Set to `false` to not select the resource of the event when clicking an event bar.
     */
    selectResourceOnEventNavigate? : boolean
    /**
     * Set to `false` to not select the row/resource when clicking the empty area in a time axis cell.
     */
    selectResourceOnScheduleClick? : boolean
    /**
     * *Only valid if this Widget is [floating](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-floating).*
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-showAnimation)
     */
    showAnimation? : boolean|object
    /**
     * Set to `false` to hide recurring fields in event editor, even if the
     * [Recurring Events](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/RecurringEvents#config-enableRecurringEvents) is `true`
     * and a recurring event is being edited.
     */
    showRecurringUI? : boolean
    /**
     * Set to `false` to not show the tooltip when this widget is [disabled](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-disabled)
     */
    showTooltipWhenDisabled? : boolean
    /**
     * An optional config object to configure the [Sidebar](https://bryntum.com/products/calendar/docs/api/Calendar/widget/Sidebar) which is
     * shown next to the calendar views specified by the [modes](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-modes).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-sidebar)
     * @param {'left'|'right'} side Which [side](https://bryntum.com/products/calendar/docs/api/Calendar/widget/Sidebar#config-side) to dock to.
     * @param {boolean} collapsed May be initially [collapsed](https://bryntum.com/products/calendar/docs/api/Calendar/widget/Sidebar#config-collapsed).
     * @param {<String, ContainerItemConfig>} items Reconfigure or add to the sidebar UI.
     * @param {ContainerItemConfig} items.custom Settings specific to custom item.
     * @param {DatePickerConfig} items.datePicker The [DatePicker](https://bryntum.com/products/calendar/docs/api/Core/widget/DatePicker) may be reconfigured using an object, or configured away using `null`. Its default [weight](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-weight) is 100.
     * @param {TextFieldConfig} items.eventFilter The eventFilter is a [TextField](https://bryntum.com/products/calendar/docs/api/Core/widget/TextField) which may be used to filter the events by matched name. It may be reconfigured using an object, or configured away using `null`. Its default [weight](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-weight) is 150.
     * @param {ResourceFilterConfig} items.resourceFilter The [ResourceFilter](https://bryntum.com/products/calendar/docs/api/Scheduler/widget/ResourceFilter) may be reconfigured using an object, or configured away using `null`. Its default [weight](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-weight) is 200.
     */
    sidebar? : {side?: 'left'|'right', collapsed?: boolean, items: { [mode: string]: ContainerItemConfig|null|undefined, datePicker?: DatePickerConfig, eventFilter?: TextFieldConfig, resourceFilter?: ResourceFilterConfig }}|boolean
    /**
     * This value can be one of the following:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-stateful)
     */
    stateful? : boolean|object|string[]
    /**
     * The events that, when fired by this component, should trigger it to save its state by calling
     * [saveState](https://bryntum.com/products/calendar/docs/api/Core/mixin/State#function-saveState).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-statefulEvents)
     */
    statefulEvents? : object|string[]
    /**
     * The key to use when saving this object's state in the [stateProvider](https://bryntum.com/products/calendar/docs/api/Core/mixin/State#config-stateProvider). If this config is
     * not assigned, and [stateful](https://bryntum.com/products/calendar/docs/api/Core/mixin/State#config-stateful) is not set to `false`, the [id](#Core/widget/Widget#config-id)
     * (if explicitly specified) will be used as the `stateId`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-stateId)
     */
    stateId? : string
    /**
     * The `StateProvider` to use to save and restore this object's [state](https://bryntum.com/products/calendar/docs/api/Core/mixin/State#property-state). By default, `state`
     * will be saved using the [default state provider](https://bryntum.com/products/calendar/docs/api/Core/state/StateProvider#property-instance-static).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-stateProvider)
     */
    stateProvider? : StateProvider
    /**
     * Specify `true` to match fields by their `name` property only when assigning a [record](https://bryntum.com/products/calendar/docs/api/Core/widget/Container#config-record),
     * without falling back to `ref`.
     */
    strictRecordMapping? : boolean
    /**
     * An object containing widgets keyed by name. By default (when no `type` is given), strips are
     * [toolbars](https://bryntum.com/products/calendar/docs/api/Core/widget/Toolbar). If you want to pass an array, you can use
     * the toolbar's [items](https://bryntum.com/products/calendar/docs/api/Core/widget/Container#config-items).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-strips)
     */
    strips? : Record<string, ContainerItemConfig>
    /**
     * A [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object, or a message to be shown when Crud Manager
     * is persisting changes on the server. Set to `null` to disable default sync mask.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-syncMask)
     */
    syncMask? : string|MaskConfig|null
    /**
     * A configuration for the [tab](https://bryntum.com/products/calendar/docs/api/Core/widget/Tab) created for this widget when it is placed in a
     * [TabPanel](https://bryntum.com/products/calendar/docs/api/Core/widget/TabPanel). For example, this config can be used to control the icon of the `tab` for
     * this widget:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-tab)
     */
    tab? : boolean|TabConfig
    /**
     * The tag name of this Widget's root element
     */
    tag? : string
    /**
     * A config object containing definitions of the toolbar of the Calendar.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-tbar)
     */
    tbar? : ToolbarConfig
    /**
     * Text alignment: 'left', 'center' or 'right'. Also accepts direction neutral 'start' and 'end'.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-textAlign)
     */
    textAlign? : 'left'|'center'|'right'|'start'|'end'
    /**
     * Inline time ranges, will be loaded into the [timeRangeStore](https://bryntum.com/products/calendar/docs/api/Calendar/mixin/CalendarStores#property-timeRangeStore).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-timeRanges)
     */
    timeRanges? : TimeRangeModel[]|TimeRangeModelConfig[]
    /**
     * The [TimeRangeStore](https://bryntum.com/products/calendar/docs/api/Calendar/data/TimeRangeStore) or a reconfiguring object for a store which will hold time ranges.
     * These may be used by the [TimeRanges](https://bryntum.com/products/calendar/docs/api/Calendar/feature/TimeRanges) feature.
     */
    timeRangeStore? : TimeRangeStore|TimeRangeStoreConfig
    /**
     * Set to a time zone or a UTC offset. This will set the projects
     * [timeZone](https://bryntum.com/products/calendar/docs/api/Scheduler/model/ProjectModel#config-timeZone) config accordingly. As this config is only a referer,
     * please se project's config [documentation](https://bryntum.com/products/calendar/docs/api/Scheduler/model/ProjectModel#config-timeZone) for more
     * information.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-timeZone)
     */
    timeZone? : string|number
    /**
     * A title to display in the header or owning TabPanel. Causes creation and docking of a header
     * to the top if no header is configured.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-title)
     */
    title? : string
    /**
     * The [tools](https://bryntum.com/products/calendar/docs/api/Core/widget/Tool) to add either before or after the `title` in the Panel header. Each
     * property name is the reference by which an instantiated tool may be retrieved from the live
     * `[tools](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Toolable#property-tools)` property.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-tools)
     */
    tools? : Record<string, ToolConfig>|null
    /**
     * Tooltip for the widget, either as a string or as a Tooltip config object.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-tooltip)
     */
    tooltip? : string|TooltipConfig|null
    /**
     * By default, tabbing within a Panel is not contained, ie you can TAB out of the Panel
     * forwards or backwards.
     * Configure this as `true` to disallow tabbing out of the Panel, and make tabbing circular within this Panel.
     */
    trapFocus? : boolean
    /**
     * Configure as `true` to trigger `selectionChange` when removing a selected event/assignment.
     */
    triggerSelectionChangeOnRemove? : boolean
    type? : 'calendar'
    /**
     * Custom CSS class name suffixes to apply to the elements rendered by this widget. This may be specified
     * as a space separated string, an array of strings, or as an object in which property names with truthy
     * values are used as the class names.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-ui)
     */
    ui? : 'plain'|'toolbar'|string|object
    /**
     * By default, the day of week and week of month of the event's start date are used to create
     * helpful contextual recurrence types when editing a non-recurring event record and choosing
     * an initial recurrence type.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-useContextualRecurrenceRules)
     */
    useContextualRecurrenceRules? : boolean
    /**
     * Configuration options to change how the subordinate [Container](https://bryntum.com/products/calendar/docs/api/Core/widget/Container) which contains
     * the calendar's child views is created.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-viewContainer)
     */
    viewContainer? : ContainerConfig
    /**
     * The week start day to be used throughout this Calendar, 0 meaning Sunday, 6 meaning Saturday.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-weekStartDay)
     */
    weekStartDay? : number
    /**
     * A widgets weight determines its position among siblings when added to a [Container](https://bryntum.com/products/calendar/docs/api/Core/widget/Container).
     * Higher weights go further down.
     */
    weight? : number
    /**
     * Widget's width, used to set element style.width. Either specify a valid width string or a number, which
     * will get 'px' appended. We recommend using CSS as the primary way to control width, but in some cases
     * this config is convenient.
     */
    width? : string|number
    /**
     * The x position for the widget.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-x)
     */
    x? : number
    /**
     * The y position for the widget.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#config-y)
     */
    y? : number

    // Events
    /**
     * Fired when a change of view has completed. By default, view changes are animated and this
     * event fires when the view is fully visible.
     * @param {object} event Event object
     * @param {Calendar} event.source This Calendar instance.
     * @param {Widget} event.prevActiveItem The previously active view.
     * @param {Widget} event.activeItem The new active view.
     */
    onActiveItemChange? : ((event: { source: Calendar, prevActiveItem: Widget, activeItem: Widget }) => void)|string
    /**
     * Fires on the owning Scheduler after editor is closed by any action - save, delete or cancel
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler
     */
    onAfterEventEdit? : ((event: { source: Scheduler }) => void)|string
    /**
     * Fires on the owning Scheduler after an event is successfully saved
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler instance
     * @param {EventModel} event.eventRecord The record about to be saved
     */
    onAfterEventSave? : ((event: { source: Scheduler, eventRecord: EventModel }) => void)|string
    /**
     * Fired before a change of view is initiated.
     * @param {object} event Event object
     * @param {Calendar} event.source This Calendar instance.
     * @param {Widget} event.prevActiveItem The previously active view.
     * @param {Widget} event.activeItem The new active view.
     */
    onBeforeActiveItemChange? : ((event: { source: Calendar, prevActiveItem: Widget, activeItem: Widget }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires before an assignment is removed. Can be triggered by user pressing [DELETE] or [BACKSPACE] or
     * by the event editor. Can for example be used to display a custom dialog to confirm deletion, in which
     * case records should be "manually" removed after confirmation:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-beforeAssignmentDelete)
     * @param {object} event Event object
     * @param {Scheduler} event.source The Scheduler instance
     * @param {EventModel[]} event.eventRecords The records about to be deleted
     * @param {object} event.context Additional removal context:
     * @param {Function} event.context.finalize Function to call to finalize the removal. Used to asynchronously decide to remove the records or not. Provide `false` to the function to prevent the removal.
     * @param {boolean} event.context.finalize.removeRecords Provide `false` to the function to prevent the removal.
     */
    onBeforeAssignmentDelete? : ((event: { source: Scheduler, eventRecords: EventModel[], context: { finalize: { removeRecords?: boolean } } }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired any time there is going to be a change to the assignments selected in the Scheduler.
     * Returning `false` prevents the change
     * @param {object} event Event object
     * @param {string} event.action One of the actions 'select', 'deselect', 'update', 'clear'
     * @param {EventModel[]} event.selected An array of assignments that will be added to the selection.
     * @param {EventModel[]} event.deselected An array of assignments that will be removed from the selection.
     * @param {EventModel[]} event.selection The currently selected assignments, before applying `selected` and `deselected`.
     */
    onBeforeAssignmentSelectionChange? : ((event: { action: string, selected: EventModel[], deselected: EventModel[], selection: EventModel[] }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires whenever a child view's [autoCreate gesture](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/CalendarMixin#config-autoCreate) is detected and also when a [drag-create](#Calendar/feature/CalendarDrag) gesture is detected.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-beforeAutoCreate)
     * @param {object} event Event object
     * @param {Event} event.domEvent The DOM event which initiated the creation.
     * @param {Date} event.date The starting time of the event to be created.
     * @param {ResourceModel} event.resourceRecord The resource if the UI includes a resource.
     * @param {typeof CalendarMixin} event.view The view in which the event is being created.
     */
    onBeforeAutoCreate? : ((event: { domEvent: Event, date: Date, resourceRecord: ResourceModel, view: typeof CalendarMixin }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Scheduler before a copy action is performed, return `false` to prevent the action
     * @param {object} event Event object
     * @param {Scheduler} event.source Owner scheduler
     * @param {EventModel[]} event.eventRecords The event records about to be copied
     * @param {AssignmentModel[]} event.assignmentRecords The assignment records about to be copied
     * @param {boolean} event.isCut `true` if this is a cut action
     * @param {string} event.entityName 'event' to distinguish this event from other beforeCopy events
     */
    onBeforeCopy? : ((event: { source: Scheduler, eventRecords: EventModel[], assignmentRecords: AssignmentModel[], isCut: boolean, entityName: string }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires before a mode is created. The configuration block is passed to the handler for possible modification.
     * @param {object} event Event object
     * @param {Calendar} event.source This Calendar
     * @param {AgendaViewConfig|DayResourceViewConfig|DayViewConfig|EventListConfig|MonthViewConfig|ResourceViewConfig|WeekViewConfig|YearViewConfig} event.config The configuration block of the mode about to be created. Any property, including the `type` may be changed.
     * @param {string} event.config.type The type of the mode about to be created.
     */
    onBeforeCreateMode? : ((event: { source: Calendar, config: { type: string } }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before CSV export starts. Return `false` to cancel the export.
     * @param {object} event Event object
     * @param {ExportConfig} event.config Export config
     * @param {Column[]} event.columns An array of columns to export
     * @param {Model[]} event.rows An array of records to export
     * @param {string} event.lineDelimiter The CSV delimiter to separate lines
     * @param {string} event.columnDelimiter The CSV delimiter to separate values on one line
     */
    onBeforeCSVExport? : ((event: { config: ExportConfig, columns: Column[], rows: Model[], lineDelimiter: string, columnDelimiter: string }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires before an object is destroyed.
     * @param {object} event Event object
     * @param {Base} event.source The Object that is being destroyed.
     */
    onBeforeDestroy? : ((event: { source: Base }) => void)|string
    /**
     * This event fires on the owning Calendar before a drag creation gesture is started. Return `false` to
     * veto the operation.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance that fired the event.
     * @param {DragContext} event.drag The drag create context.
     * @param {Event} event.domEvent The browser event.
     * @param {Date} event.date The date at the drag DOM event position.
     * @param {ResourceModel} event.resourceRecord The `ResourceModel` record if the gesture was performed in a resource-type view.
     * @param {CalendarDrag} event.feature The Calendar drag feature instance.
     * @param {Widget} event.view The Calendar widget in which the drag is being started.
     */
    onBeforeDragCreate? : ((event: { source: Calendar, drag: DragContext, domEvent: Event, date: Date, resourceRecord?: ResourceModel, feature: CalendarDrag, view: Widget }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Calendar before a drag creation gesture is completed. Return `false` to
     * immediately veto the operation or a Promise yielding `true` or `false` for async vetoing.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance that fired the event.
     * @param {DragContext} event.drag The drag create context.
     * @param {Event} event.event The browser event.
     * @param {EventModel} event.eventRecord The `EventModel` record being created that has not yet been added in the store.
     * @param {Date} event.newStartDate The new start date.
     * @param {Date} event.newEndDate The new end date.
     * @param {ResourceModel} event.resourceRecord The `ResourceModel` record if the gesture was performed in a resource-type view.
     * @param {CalendarDrag} event.feature The Calendar drag feature instance.
     * @param {boolean|ValidateCreateResult} event.validation The result of the [validateCreateFn](https://bryntum.com/products/calendar/docs/api/Calendar/feature/CalendarDrag#config-validateCreateFn) if one was provided.
     * @param {Widget} event.view The Calendar widget in which the drag completed.
     */
    onBeforeDragCreateEnd? : ((event: { source: Calendar, drag: DragContext, event: Event, eventRecord: EventModel, newStartDate: Date, newEndDate: Date, resourceRecord?: ResourceModel, feature: CalendarDrag, validation: boolean|ValidateCreateResult, view: Widget }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Calendar before a drag move gesture is started. Return `false` to
     * veto the operation.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance that fired the event.
     * @param {DragContext} event.drag The drag move context.
     * @param {Event} event.domEvent The browser event.
     * @param {EventModel} event.eventRecord The `EventModel` record being moved.
     * @param {Date} event.date The date at the drag DOM event position.
     * @param {ResourceModel} event.resourceRecord The `ResourceModel` record if the gesture was performed in a resource-type view.
     * @param {CalendarDrag} event.feature The Calendar drag feature instance.
     * @param {Widget} event.view The Calendar widget in which the drag is being started.
     */
    onBeforeDragMove? : ((event: { source: Calendar, drag: DragContext, domEvent: Event, eventRecord: EventModel, date: Date, resourceRecord?: ResourceModel, feature: CalendarDrag, view: Widget }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Calendar before a drag move gesture is completed. Return `false` to immediately veto the operation
     * or a Promise yielding `true` or `false` for async vetoing.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance that fired the event.
     * @param {DragContext} event.drag The drag create context.
     * @param {Event} event.event The browser event.
     * @param {EventModel} event.eventRecord The `EventModel` record that has not yet been updated in the store.
     * @param {Date} event.newStartDate The new start date.
     * @param {Date} event.newEndDate The new end date.
     * @param {ResourceModel} event.resourceRecord The `ResourceModel` record if the gesture was performed in a resource-type view.
     * @param {CalendarDrag} event.feature The Calendar drag feature instance.
     * @param {boolean|ValidateCreateResult} event.validation The result of the [validateMoveFn](https://bryntum.com/products/calendar/docs/api/Calendar/feature/CalendarDrag#config-validateMoveFn) if one was provided.
     * @param {Widget} event.view The Calendar widget in which the drag completed.
     */
    onBeforeDragMoveEnd? : ((event: { source: Calendar, drag: DragContext, event: Event, eventRecord: EventModel, newStartDate: Date, newEndDate: Date, resourceRecord?: ResourceModel, feature: CalendarDrag, validation: boolean|ValidateCreateResult, view: Widget }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Calendar before a drag resize gesture is started. Return `false` to
     * veto the operation.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance that fired the event.
     * @param {DragContext} event.drag The drag resize context.
     * @param {Event} event.domEvent The browser event.
     * @param {EventModel} event.eventRecord The `EventModel` record being resized.
     * @param {Date} event.date The date at the drag DOM event position.
     * @param {ResourceModel} event.resourceRecord The `ResourceModel` record if the gesture was performed in a resource-type view.
     * @param {CalendarDrag} event.feature The Calendar drag feature instance.
     * @param {Widget} event.view The Calendar widget in which the drag is being started.
     */
    onBeforeDragResize? : ((event: { source: Calendar, drag: DragContext, domEvent: Event, eventRecord: EventModel, date: Date, resourceRecord?: ResourceModel, feature: CalendarDrag, view: Widget }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Calendar before a drag resize gesture is completed. Return `false` to immediately veto the operation
     * or a Promise yielding `true` or `false` for async vetoing.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance that fired the event.
     * @param {DragContext} event.drag The drag create context.
     * @param {Event} event.event The browser event.
     * @param {EventModel} event.eventRecord The `EventModel` record that has not yet been updated in the store.
     * @param {Date} event.newStartDate The new start date.
     * @param {Date} event.newEndDate The new end date.
     * @param {CalendarDrag} event.feature The Calendar drag feature instance.
     * @param {boolean|ValidateCreateResult} event.validation The result of the [validateResizeFn](https://bryntum.com/products/calendar/docs/api/Calendar/feature/CalendarDrag#config-validateResizeFn) if one was provided.
     * @param {Widget} event.view The Calendar widget in which the drag completed.
     */
    onBeforeDragResizeEnd? : ((event: { source: Calendar, drag: DragContext, event: Event, eventRecord: EventModel, newStartDate: Date, newEndDate: Date, feature: CalendarDrag, validation: boolean|ValidateCreateResult, view: Widget }) => Promise<boolean>|boolean|void)|string
    /**
     * This event is fired on the owning Calendar when dropping an event from the calendar on the
     * external source <strong>if the [droppable](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-droppable) was configured</strong>. Returning `false`
     * prevents the gesture from being completed.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-beforeDropExternal)
     * @param {object} event Event object
     * @param {EventModel} event.eventRecord The event record being dragged.
     * @param {boolean} event.dropOnCalendar `true` if the drop gesture is over the client Calendar. This feature also allows drag *out* of the Calendar and onto the external event source if the [droppable](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-droppable) config is set.
     * @param {HTMLElement} event.itemElement The element in which the drag gesture started.
     * @param {HTMLElement} event.targetElement The current over element.
     * @param {Event} event.domEvent The pointer event associated with the drag point.
     * @param {number} event.overIndex *If [grid](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-grid) was specified*, the index of the row being moved over;
     * @param {Model} event.overRecord *If [grid](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-grid) was specified*, the record being moved over;
     * @param {boolean} event.isAbove *If [grid](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-grid) was specified*, `true` if the pointer position is above the halfway line of the over row.
     * @param {boolean} event.altKey `true` if the Alt key was down when the last event was processed.
     * @param {boolean} event.ctrlKey `true` if the Ctrl key was down when the last event was processed.
     * @param {boolean} event.metaKey `true` if the Meta key was down when the last event was processed.
     * @param {boolean} event.shiftKey `true` if the Shift key was down when the last event was processed.
     */
    onBeforeDropExternal? : ((event: { eventRecord: EventModel, dropOnCalendar: boolean, itemElement: HTMLElement, targetElement: HTMLElement, domEvent: Event, overIndex: number, overRecord: Model, isAbove: boolean, altKey: boolean, ctrlKey: boolean, metaKey: boolean, shiftKey: boolean }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires before an event is removed. Can be triggered by user pressing [DELETE] or [BACKSPACE] or by the
     * event editor. Can for example be used to display a custom dialog to confirm deletion, in which case
     * records should be "manually" removed after confirmation:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-beforeEventDelete)
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance
     * @param {EventModel[]} event.eventRecords The records about to be deleted
     * @param {object} event.context Additional removal context:
     * @param {Function} event.context.finalize Function to call to finalize the removal. Used to asynchronously decide to remove the records or not. Provide `false` to the function to prevent the removal.
     * @param {boolean} event.context.finalize.removeRecords Provide `false` to the function to prevent the removal.
     */
    onBeforeEventDelete? : ((event: { source: Calendar, eventRecords: EventModel[], context: { finalize: { removeRecords?: boolean } } }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Scheduler before an event is displayed in an editor.
     * This may be listened for to allow an application to take over event editing duties. Returning `false`
     * stops the default editing UI from being shown.
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler
     * @param {SchedulerEventEdit} event.eventEdit The eventEdit feature
     * @param {EventModel} event.eventRecord The record about to be shown in the event editor.
     * @param {ResourceModel} event.resourceRecord The Resource record for the event. If the event is being created, it will not contain a resource, so this parameter specifies the resource the event is being created for.
     * @param {HTMLElement} event.eventElement The element which represents the event in the scheduler display.
     */
    onBeforeEventEdit? : ((event: { source: Scheduler, eventEdit: SchedulerEventEdit, eventRecord: EventModel, resourceRecord: ResourceModel, eventElement: HTMLElement }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Scheduler when the editor for an event is available but before it is populated with
     * data and shown. Allows manipulating fields etc.
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler
     * @param {SchedulerEventEdit} event.eventEdit The eventEdit feature
     * @param {EventModel} event.eventRecord The record about to be shown in the event editor.
     * @param {ResourceModel} event.resourceRecord The Resource record for the event. If the event is being created, it will not contain a resource, so this parameter specifies the resource the event is being created for.
     * @param {HTMLElement} event.eventElement The element which represents the event in the scheduler display.
     * @param {Popup} event.editor The editor
     */
    onBeforeEventEditShow? : ((event: { source: Scheduler, eventEdit: SchedulerEventEdit, eventRecord: EventModel, resourceRecord: ResourceModel, eventElement: HTMLElement, editor: Popup }) => void)|string
    /**
     * Fires on the owning Scheduler before an event is saved.
     * Return `false` to immediately prevent saving
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-beforeEventSave)
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler instance
     * @param {EventModel} event.eventRecord The record about to be saved
     * @param {ResourceModel[]} event.resourceRecords The resources to which the event is assigned
     * @param {object} event.values The new values
     * @param {object} event.context Extended save context:
     * @param {boolean} event.context.async Set this to `true` in a listener to indicate that the listener will asynchronously decide to prevent or not the event save.
     * @param {Function} event.context.finalize Function to call to finalize the save. Used when `async` is `true`. Provide `false` to the function to prevent the save.
     */
    onBeforeEventSave? : ((event: { source: Scheduler, eventRecord: EventModel, resourceRecords: ResourceModel[], values: object, context: { async?: boolean, finalize: Function } }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired any time there is going to be a change to the events selected in the Scheduler.
     * Returning `false` prevents the change
     * @param {object} event Event object
     * @param {string} event.action One of the actions 'select', 'deselect', 'update', 'clear'
     * @param {EventModel[]} event.selected An array of events that will be added to the selection.
     * @param {EventModel[]} event.deselected An array of events that will be removed from the selection.
     * @param {EventModel[]} event.selection The currently selected events, before applying `selected` and `deselected`.
     */
    onBeforeEventSelectionChange? : ((event: { action: string, selected: EventModel[], deselected: EventModel[], selection: EventModel[] }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before Excel export starts. Return `false` to cancel the export.
     * @param {object} event Event object
     * @param {ExportConfig} event.config Export config
     * @param {XLSColumn[]} event.columns An array of columns to export
     * @param {any[]} event.rows An array of records to export
     */
    onBeforeExcelExport? : ((event: { config: ExportConfig, columns: XLSColumn[], rows: any[] }) => Promise<boolean>|boolean|void)|string
    /**
     * Triggered before a widget is hidden. Return `false` to prevent the action.
     * @param {object} event Event object
     * @param {Widget} event.source The widget being hidden.
     */
    onBeforeHide? : ((event: { source: Widget }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Scheduler before a paste action is performed, return `false` to prevent the action.
     * Depending on if the EventStore is using [singleAssignment](https://bryntum.com/products/calendar/docs/api/Scheduler/data/EventStore#config-singleAssignment) or the
     * configuration of [copyPasteAction](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/EventCopyPaste#config-copyPasteAction), either the `eventRecords` or the `assignmentRecords` param
     * will be populated with record copies.
     * @param {object} event Event object
     * @param {Scheduler} event.source Owner scheduler
     * @param {EventModel[]} event.eventRecords The events about to be pasted
     * @param {AssignmentModel[]} event.assignmentRecords The assignments about to be pasted
     * @param {EventModel[]} event.originalEventRecords The original event records
     * @param {AssignmentModel[]} event.originalAssignmentRecords The original assignment record
     * @param {Date} event.date The date when the pasted events will be scheduled
     * @param {ResourceModel} event.resourceRecord The target resource record, the clipboard event records will be assigned to this resource.
     * @param {boolean} event.isCut `true` if this is a cut action
     * @param {string} event.entityName 'event' to distinguish this event from other beforePaste events
     */
    onBeforePaste? : ((event: { source: Scheduler, eventRecords: EventModel[], assignmentRecords: AssignmentModel[], originalEventRecords: EventModel[], originalAssignmentRecords: AssignmentModel[], date: Date, resourceRecord: ResourceModel, isCut: boolean, entityName: string }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Calendar before print starts. Return `false` to cancel the print.
     * @param {object} event Event object
     * @param {object} event.config Print config
     */
    onBeforePrint? : ((event: { config: object }) => Promise<boolean>|boolean|void)|string
    /**
     * Triggered before a new [responsiveState](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Responsive#config-responsiveState) is applied.
     * @param {object} event Event object
     * @param {Widget} event.source The widget whose `responsiveState` is to be changed
     * @param {string} event.state The new value for the widget's `responsiveState`
     * @param {string} event.oldState The previous value for the widget's `responsiveState`
     */
    onBeforeResponsiveStateChange? : ((event: { source: Widget, state: string, oldState: string }) => void)|string
    /**
     * Fired before this container will load record values into its child fields. This is useful if you
     * want to modify the UI before data is loaded (e.g. set some input field to be readonly)
     * @param {object} event Event object
     * @param {Container} event.source The container
     * @param {Model} event.record The record
     */
    onBeforeSetRecord? : ((event: { source: Container, record: Model }) => void)|string
    /**
     * Triggered before a widget is shown. Return `false` to prevent the action.
     * @param {object} event Event object
     * @param {Widget|any} event.source The widget being shown
     */
    onBeforeShow? : ((event: { source: Widget|any }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired before state is applied to the source. Allows editing the state object or preventing the operation.
     * @param {object} event Event object
     * @param {object} event.event The Bryntum event object
     * @param {object} event.event.state State object config
     */
    onBeforeStateApply? : ((event: { event: { state: object } }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired before state is saved by the StateProvider. Allows editing the state object or preventing the operation.
     * @param {object} event Event object
     * @param {object} event.event The Bryntum event object
     * @param {object} event.event.state State object config
     */
    onBeforeStateSave? : ((event: { event: { state: object } }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires when any other event is fired from the object.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-catchAll)
     * @param {object} event Event object
     * @param {{[key: string]: any, type: string}} event.event The Object that contains event details
     * @param {string} event.event.type The type of the event which is caught by the listener
     */
    onCatchAll? : ((event: {[key: string]: any, type: string}) => void)|string
    /**
     * Fires when a cell overflow indicator is clicked in any view. This will include
     * the cells of a [YearView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/YearView) and the `+ n more` overflow
     * indicator of cells in a [MonthView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/MonthView), and "all day" events section
     * of a [DayView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView).
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date clicked on.
     */
    onCellOverflowClick? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date }) => void)|string
    /**
     * Fires when a Panel is collapsed using the [collapsible](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-collapsible) setting.
     * @param {object} event Event object
     * @param {Panel} event.source This Panel.
     */
    onCollapse? : ((event: { source: Panel }) => void)|string
    /**
     * This event fires on the owning widget when an item is selected in the context menu.
     * @param {object} event Event object
     * @param {Widget} event.source The owning widget
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     */
    onContextMenuItem? : ((event: { source: Widget, menu: Menu, item: MenuItem }) => void)|string
    /**
     * This event fires on the owning widget when a check item is toggled in the context menu.
     * @param {object} event Event object
     * @param {Widget} event.source The owning widget
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     * @param {boolean} event.checked Checked or not
     */
    onContextMenuToggleItem? : ((event: { source: Widget, menu: Menu, item: MenuItem, checked: boolean }) => void)|string
    /**
     * Fires on the owning Scheduler after a copy action is performed. Depending on if the EventStore is using
     * [singleAssignment](https://bryntum.com/products/calendar/docs/api/Scheduler/data/EventStore#config-singleAssignment) or the configuration of
     * [copyPasteAction](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/EventCopyPaste#config-copyPasteAction), either the `eventRecords` or the `assignmentRecords` param will be populated
     * with record copies.
     * @param {object} event Event object
     * @param {Scheduler} event.source Owner scheduler
     * @param {EventModel[]} event.eventRecords The event record copies
     * @param {AssignmentModel[]} event.assignmentRecords The assignment record copies
     * @param {EventModel[]} event.originalEventRecords The event records that were copied
     * @param {AssignmentModel[]} event.originalAssignmentRecord The assignment records that were copied
     * @param {boolean} event.isCut `true` if this is a cut action
     * @param {string} event.entityName 'event' to distinguish this event from other copy events
     */
    onCopy? : ((event: { source: Scheduler, eventRecords: EventModel[], assignmentRecords: AssignmentModel[], originalEventRecords: EventModel[], originalAssignmentRecord: AssignmentModel[], isCut: boolean, entityName: string }) => void)|string
    /**
     * Fires on the owning Scheduler/Gantt when the line indicating the current time is updated (see [currentTimeLineUpdateInterval](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/TimeRanges#config-currentTimeLineUpdateInterval)).
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler
     * @param {Date} event.date The current date
     */
    onCurrentTimelineUpdate? : ((event: { source: Scheduler, date: Date }) => void)|string
    /**
     * Fired when data in any of the projects stores changes.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-dataChange)
     * @param {object} event Event object
     * @param {typeof ProjectConsumer|any} event.source Owning component
     * @param {typeof ProjectModelMixin|any} event.project Project model
     * @param {Store} event.store Affected store
     * @param {'remove'|'removeAll'|'add'|'clearchanges'|'filter'|'update'|'dataset'|'replace'} event.action Name of action which triggered the change. May be one of:  * `'remove'` * `'removeAll'` * `'add'` * `'clearchanges'` * `'filter'` * `'update'` * `'dataset'` * `'replace'`
     * @param {Model} event.record Changed record, for actions that affects exactly one record (`'update'`)
     * @param {Model[]} event.records Changed records, passed for all actions except `'removeAll'`
     * @param {object} event.changes Passed for the `'update'` action, info on which record fields changed
     */
    onDataChange? : ((event: { source: typeof ProjectConsumer|any, project: typeof ProjectModelMixin|any, store: Store, action: 'remove'|'removeAll'|'add'|'clearchanges'|'filter'|'update'|'dataset'|'replace', record: Model, records: Model[], changes: object }) => void)|string
    /**
     * Fires when the calendar changes the date that it orientates its views around.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-dateChange)
     * @param {object} event Event object
     * @param {Calendar} event.source This Calendar.
     * @param {Date} event.oldDate The previous Calendar date.
     * @param {Date} event.date The new Calendar date.
     */
    onDateChange? : ((event: { source: Calendar, oldDate: Date, date: Date }) => void)|string
    /**
     * Fired when the eventStore is queried for events by date range, and the date range
     * requested is different from the last time the store was queried.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-dateRangeChange)
     * @param {object} event Event object
     * @param {Calendar} event.source This Calendar instance
     * @param {object} event.old The old date range
     * @param {Date} event.old.startDate the old start date
     * @param {Date} event.old.endDate the old end date
     * @param {object} event.new The new date range
     * @param {Date} event.new.startDate the new start date
     * @param {Date} event.new.endDate the new end date
     */
    onDateRangeChange? : ((event: { source: Calendar, old: { startDate: Date, endDate: Date }, new: { startDate: Date, endDate: Date } }) => void)|string
    /**
     * Fires when the [LoadOnDemand](https://bryntum.com/products/calendar/docs/api/Calendar/feature/LoadOnDemand) feature has loaded a range of events.
     * @param {object} event Event object
     * @param {object} event.response The decoded JSON response.
     * @param {object} event.options The options object passed into the CrudManager [CrudManager](https://bryntum.com/products/calendar/docs/api/Scheduler/data/CrudManager) [load](https://bryntum.com/products/calendar/docs/api/Scheduler/crud/AbstractCrudManagerMixin#function-load) method.
     * @param {Date} event.startDate The start date of the range to request.
     * @param {Date} event.endDate The end date of the range to request. <strong>Note that Dates are timestamps</strong>.
     */
    onDateRangeLoad? : ((event: { response: object, options: object, startDate: Date, endDate: Date }) => void)|string
    /**
     * Fired <strong>every</strong> time the eventStore is queried for events by date range.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-dateRangeRequested)
     * @param {object} event Event object
     * @param {Calendar} event.source This Calendar instance
     * @param {object} event.old The old date range
     * @param {Date} event.old.startDate the old start date
     * @param {Date} event.old.endDate the old end date
     * @param {object} event.new The new date range
     * @param {Date} event.new.startDate the new start date
     * @param {Date} event.new.endDate the new end date
     * @param {boolean} event.changed `true` if the date range is different from the last time a request was made
     */
    onDateRangeRequested? : ((event: { source: Calendar, old: { startDate: Date, endDate: Date }, new: { startDate: Date, endDate: Date }, changed: boolean }) => void)|string
    /**
     * Fires when a day number is clicked in a view which shows day numbers. This will include
     * the cells of a [YearView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/YearView), the cell header of cells
     * in a [MonthView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/MonthView), and the day header in a
     * [DayView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView).
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date clicked on.
     * @param {DayCell} event.cellData An object that contains data about the calendar cell for the date.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onDayNumberClick? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, cellData: DayCell, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fires when an object is destroyed.
     * @param {object} event Event object
     * @param {Base} event.source The Object that is being destroyed.
     */
    onDestroy? : ((event: { source: Base }) => void)|string
    /**
     * Fires when a field is mutated and the state of the [hasChanges](https://bryntum.com/products/calendar/docs/api/Core/widget/Container#property-hasChanges) property changes
     * @param {object} event Event object
     * @param {Container} event.source The container.
     * @param {boolean} event.dirty The dirty state of the Container - `true` if there are any fields which have been changed since initial load.
     */
    onDirtyStateChange? : ((event: { source: Container, dirty: boolean }) => void)|string
    /**
     * This event fires on the owning Calendar when a drag creation gesture is completed.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance that fired the event.
     * @param {DragContext} event.drag The drag create context.
     * @param {Event} event.event The browser event.
     * @param {EventModel} event.eventRecord The new `EventModel` record added in the store.
     * @param {ResourceModel} event.resourceRecord The `ResourceModel` record if the gesture was performed in a resource-type view.
     * @param {CalendarDrag} event.feature The Calendar drag feature instance.
     * @param {boolean|ValidateCreateResult} event.validation The result of the [validateCreateFn](https://bryntum.com/products/calendar/docs/api/Calendar/feature/CalendarDrag#config-validateCreateFn) if one was provided.
     * @param {Widget} event.view The Calendar widget in which the drag completed.
     */
    onDragCreateEnd? : ((event: { source: Calendar, drag: DragContext, event: Event, eventRecord: EventModel, resourceRecord?: ResourceModel, feature: CalendarDrag, validation: boolean|ValidateCreateResult, view: Widget }) => void)|string
    /**
     * This event fires on the owning Calendar when a drag move gesture is completed. The `eventRecord` has already been added
     * to the `eventStore` of the owning calendar.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance that fired the event.
     * @param {DragContext} event.drag The drag create context.
     * @param {Event} event.event The browser event.
     * @param {EventModel} event.eventRecord The updated `EventModel` record.
     * @param {ResourceModel} event.resourceRecord The `ResourceModel` record if the gesture was performed in a resource-type view.
     * @param {CalendarDrag} event.feature The Calendar drag feature instance.
     * @param {boolean|ValidateCreateResult} event.validation The result of the [validateMoveFn](https://bryntum.com/products/calendar/docs/api/Calendar/feature/CalendarDrag#config-validateMoveFn) if one was provided.
     * @param {Widget} event.view The Calendar widget in which the drag completed.
     */
    onDragMoveEnd? : ((event: { source: Calendar, drag: DragContext, event: Event, eventRecord: EventModel, resourceRecord?: ResourceModel, feature: CalendarDrag, validation: boolean|ValidateCreateResult, view: Widget }) => void)|string
    /**
     * This event is fired on the owning Calendar when dragging an event from the calendar over the
     * external source <strong>if the [droppable](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-droppable) was configured</strong>.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-dragMoveExternal)
     * @param {object} event Event object
     * @param {EventModel} event.eventRecord The event record being dragged.
     * @param {HTMLElement} event.itemElement The element in which the drag gesture started.
     * @param {HTMLElement} event.targetElement The current over element.
     * @param {Event} event.domEvent The pointer event associated with the drag point.
     * @param {number} event.overIndex *If [grid](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-grid) was specified*, the index of the row being moved over;
     * @param {Model} event.overRecord *If [grid](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-grid) was specified*, the record being moved over;
     * @param {boolean} event.isAbove *If [grid](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-grid) was specified*, `true` if the pointer position is above the halfway line of the over row.
     * @param {boolean} event.altKey `true` if the Alt key was down when the last event was processed.
     * @param {boolean} event.ctrlKey `true` if the Ctrl key was down when the last event was processed.
     * @param {boolean} event.metaKey `true` if the Meta key was down when the last event was processed.
     * @param {boolean} event.shiftKey `true` if the Shift key was down when the last event was processed.
     */
    onDragMoveExternal? : ((event: { eventRecord: EventModel, itemElement: HTMLElement, targetElement: HTMLElement, domEvent: Event, overIndex: number, overRecord: Model, isAbove: boolean, altKey: boolean, ctrlKey: boolean, metaKey: boolean, shiftKey: boolean }) => void)|string
    /**
     * This event fires on the owning Calendar when a drag resize gesture is completed.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance that fired the event.
     * @param {DragContext} event.drag The drag create context.
     * @param {Event} event.event The browser event.
     * @param {EventModel} event.eventRecord The updated `EventModel` record.
     * @param {CalendarDrag} event.feature The Calendar drag feature instance.
     * @param {boolean|ValidateCreateResult} event.validation The result of the [validateResizeFn](https://bryntum.com/products/calendar/docs/api/Calendar/feature/CalendarDrag#config-validateResizeFn) if one was provided.
     * @param {Widget} event.view The Calendar widget in which the drag completed.
     */
    onDragResizeEnd? : ((event: { source: Calendar, drag: DragContext, event: Event, eventRecord: EventModel, feature: CalendarDrag, validation: boolean|ValidateCreateResult, view: Widget }) => void)|string
    /**
     * This event is fired on the owning Calendar after dropping an event from the calendar on the
     * external source has been completed <strong>if the [droppable](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-droppable) was configured</strong>.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-dropExternal)
     * @param {object} event Event object
     * @param {EventModel} event.eventRecord The event record that was dropped.
     * @param {boolean} event.dropOnCalendar `true` if the drop gesture is over the client Calendar. This feature also allows drag *out* of the Calendar and onto the external event source if the [droppable](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-droppable) config is set.
     * @param {HTMLElement} event.itemElement The element in which the drag gesture started.
     * @param {HTMLElement} event.targetElement The current over element.
     * @param {Event} event.domEvent The pointer event associated with the drag point.
     * @param {number} event.overIndex *If [grid](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-grid) was specified*, the index of the row being moved over;
     * @param {Model} event.overRecord *If [grid](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-grid) was specified*, the record being moved over;
     * @param {boolean} event.isAbove *If [grid](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource#config-grid) was specified*, `true` if the pointer position is above the halfway line of the over row.
     * @param {boolean} event.altKey `true` if the Alt key was down when the last event was processed.
     * @param {boolean} event.ctrlKey `true` if the Ctrl key was down when the last event was processed.
     * @param {boolean} event.metaKey `true` if the Meta key was down when the last event was processed.
     * @param {boolean} event.shiftKey `true` if the Shift key was down when the last event was processed.
     */
    onDropExternal? : ((event: { eventRecord: EventModel, dropOnCalendar: boolean, itemElement: HTMLElement, targetElement: HTMLElement, domEvent: Event, overIndex: number, overRecord: Model, isAbove: boolean, altKey: boolean, ctrlKey: boolean, metaKey: boolean, shiftKey: boolean }) => void)|string
    /**
     * Fired when an empty cell content area is clicked on in a [MonthView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/MonthView) or a
     * [CalendarRow](https://bryntum.com/products/calendar/docs/api/Calendar/widget/CalendarRow).
     * @param {object} event Event object
     * @param {Event} event.domEvent The triggering DOM event.
     * @param {Date} event.date The date which has no visible events
     */
    onEmptyCellClick? : ((event: { domEvent: Event, date: Date }) => void)|string
    /**
     * Fired when an [autoCreate](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/CalendarMixin#config-autoCreate) gesture has
     * created a new event and added it to the event store.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-eventAutoCreated)
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The child view on which the event was initiated.
     * @param {EventModel} event.eventRecord The new event record.
     */
    onEventAutoCreated? : ((event: { source: typeof CalendarMixin, eventRecord: EventModel }) => void)|string
    /**
     * Fires when an event bar is clicked in any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event clicked on.
     * @param {ResourceModel} event.resourceRecord The resource clicked on if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventClick? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fires when an event bar is right-clicked in any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event right-clicked on.
     * @param {ResourceModel} event.resourceRecord The resource right-clicked on if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventContextMenu? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fires when an event bar is double clicked in any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event double clicked on.
     * @param {ResourceModel} event.resourceRecord The resource double clicked on if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventDblClick? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fired before the editor will load the event record data into its input fields. This is useful if you
     * want to modify the fields before data is loaded (e.g. set some input field to be readonly)
     * @param {object} event Event object
     * @param {Container} event.source The editor widget
     * @param {EventModel} event.record The record
     */
    onEventEditBeforeSetRecord? : ((event: { source: Container, record: EventModel }) => void)|string
    /**
     * Fires when a key is pressed while an event bar is focused.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event in which the key event was detected.
     * @param {ResourceModel} event.resourceRecord The resource in which the key event was detected if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventKeyDown? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * This event fires on the owning Scheduler before the context menu is shown for an event. Allows manipulation of the items
     * to show in the same way as in `processItems`. Returning `false` from a listener prevents the menu from
     * being shown.
     * @param {object} event Event object
     * @param {Scheduler} event.source
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {EventModel} event.eventRecord Event record for which the menu was triggered
     * @param {ResourceModel} event.resourceRecord Resource record
     * @param {AssignmentModel} event.assignmentRecord Assignment record, if assignments are used
     * @param {HTMLElement} event.eventElement
     * @param {MouseEvent} event.event Pointer event which triggered the context menu (if any)
     */
    onEventMenuBeforeShow? : ((event: { source: Scheduler, items: Record<string, MenuItemConfig>, eventRecord: EventModel, resourceRecord: ResourceModel, assignmentRecord: AssignmentModel, eventElement: HTMLElement, event?: MouseEvent }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Scheduler when an item is selected in the context menu.
     * @param {object} event Event object
     * @param {Scheduler} event.source
     * @param {MenuItem} event.item
     * @param {EventModel} event.eventRecord
     * @param {ResourceModel} event.resourceRecord
     * @param {AssignmentModel} event.assignmentRecord Assignment record, if assignments are used
     * @param {HTMLElement} event.eventElement
     */
    onEventMenuItem? : ((event: { source: Scheduler, item: MenuItem, eventRecord: EventModel, resourceRecord: ResourceModel, assignmentRecord: AssignmentModel, eventElement: HTMLElement }) => void)|string
    /**
     * This event fires on the owning Scheduler after showing the context menu for an event
     * @param {object} event Event object
     * @param {Scheduler} event.source
     * @param {Menu} event.menu The menu
     * @param {EventModel} event.eventRecord Event record for which the menu was triggered
     * @param {ResourceModel} event.resourceRecord Resource record
     * @param {AssignmentModel} event.assignmentRecord Assignment record, if assignments are used
     * @param {HTMLElement} event.eventElement
     */
    onEventMenuShow? : ((event: { source: Scheduler, menu: Menu, eventRecord: EventModel, resourceRecord: ResourceModel, assignmentRecord: AssignmentModel, eventElement: HTMLElement }) => void)|string
    /**
     * Fires when an event bar is mousedowned in any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event mousedowned on.
     * @param {ResourceModel} event.resourceRecord The resource mousedowned on if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventMouseDown? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fires when cursor enters an event bar in any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event record.
     * @param {ResourceModel} event.resourceRecord The resource if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventMouseEnter? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fires when cursor leaves an event bar in any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event record.
     * @param {ResourceModel} event.resourceRecord The resource if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventMouseLeave? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fires when an event bar is mouseouted in any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event mouseouted.
     * @param {ResourceModel} event.resourceRecord The resource mouseouted if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventMouseOut? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fires when an event bar is mouseovered in any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event mouseovered.
     * @param {ResourceModel} event.resourceRecord The resource mouseovered if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventMouseOver? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fires when an event bar is mouseupped in any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date of the UI element which contains the event.
     * @param {HTMLElement} event.eventElement The UI element which represents the event.
     * @param {EventModel} event.eventRecord The event mouseupped.
     * @param {ResourceModel} event.resourceRecord The resource mouseupped if the UI includes a resource.
     * @param {boolean} event.fromOverflowPopup `true` if the interaction was through the source view's [overflowPopup](https://bryntum.com/products/calendar/docs/api/Calendar/widget/mixin/DayCellRenderer#property-overflowPopup).
     */
    onEventMouseUp? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date, eventElement: HTMLElement, eventRecord: EventModel, resourceRecord: ResourceModel, fromOverflowPopup: boolean }) => void)|string
    /**
     * Fires when a day spanning event is found, and the date to which its encapsulating event bar
     * extends has been calculated.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-eventPropagate)
     * @param {object} event Event object
     * @param {Date} event.eventEndDate The end date for which to calculate the propagate end date.
     * @param {Date} event.propagateEndDate The system-calculated end point of the event bar.
     * @param {boolean} event.isAllDay `true` if the event is an all day event, or spans multiple days.
     * @param {boolean} event.isOverflow `true` if this is being called as part of further propagation.
     * @param {boolean} event.overflows `true` if the event extends into future cells.
     * @param {EventModel} event.eventRecord The event record being propagated.
     * @param {Date} event.date The date from which the event is being propagated.
     */
    onEventPropagate? : ((event: { eventEndDate: Date, propagateEndDate: Date, isAllDay: boolean, isOverflow: boolean, overflows: boolean, eventRecord: EventModel, date: Date }) => void)|string
    /**
     * Fires when a Panel is expanded using the [collapsible](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-collapsible) setting.
     * @param {object} event Event object
     * @param {Panel} event.source This Panel.
     */
    onExpand? : ((event: { source: Panel }) => void)|string
    /**
     * Fired when focus enters this Widget.
     * @param {object} event Event object
     * @param {Widget} event.source This Widget
     * @param {HTMLElement} event.fromElement The element which lost focus.
     * @param {HTMLElement} event.toElement The element which gained focus.
     * @param {Widget} event.fromWidget The widget which lost focus.
     * @param {Widget} event.toWidget The widget which gained focus.
     * @param {boolean} event.backwards `true` if the `toElement` is before the `fromElement` in document order.
     */
    onFocusIn? : ((event: { source: Widget, fromElement: HTMLElement, toElement: HTMLElement, fromWidget: Widget, toWidget: Widget, backwards: boolean }) => void)|string
    /**
     * Fired when focus exits this Widget's ownership tree. This is different from a `blur` event.
     * focus moving from within this Widget's ownership tree, even if there are floating widgets
     * will not trigger this event. This is when focus exits this widget completely.
     * @param {object} event Event object
     * @param {Widget} event.source This Widget
     * @param {HTMLElement} event.fromElement The element which lost focus.
     * @param {HTMLElement} event.toElement The element which gained focus.
     * @param {Widget} event.fromWidget The widget which lost focus.
     * @param {Widget} event.toWidget The widget which gained focus.
     * @param {boolean} event.backwards `true` if the `toElement` is before the `fromElement` in document order.
     */
    onFocusOut? : ((event: { source: Widget, fromElement: HTMLElement, toElement: HTMLElement, fromWidget: Widget, toWidget: Widget, backwards: boolean }) => void)|string
    /**
     * Triggered after a widget was hidden
     * @param {object} event Event object
     * @param {Widget} event.source The widget
     */
    onHide? : ((event: { source: Widget }) => void)|string
    /**
     * Fires when the [LoadOnDemand](https://bryntum.com/products/calendar/docs/api/Calendar/feature/LoadOnDemand) feature detects that a request
     * for data from the server has failed.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-loadOnDemandFail)
     * @param {object} event Event object
     * @param {any} event.rawResponse The HTTP `fetch` response object.
     * @param {object} event.request The CrudManager load data block.
     * @param {object} event.response The decoded JSON response.
     */
    onLoadOnDemandFail? : ((event: { rawResponse: any, request: object, response: object }) => void)|string
    /**
     * Fires when a month name header in a [YearView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/YearView) is clicked on.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {number} event.month The month index *(zero based)* clicked on.
     * @param {Date} event.date The date of the first of the month clicked on.
     */
    onMonthNameClick? : ((event: { source: typeof CalendarMixin, domEvent: Event, month: number, date: Date }) => void)|string
    /**
     * Fired when a user gesture causes the active item to change.
     * @param {object} event Event object
     * @param {Event} event.event The browser event which instigated navigation. May be a click or key or focus event.
     * @param {HTMLElement|null} event.item The newly active item, or `null` if focus moved out.
     * @param {HTMLElement|null} event.oldItem The previously active item, or `null` if focus is moving in.
     */
    onNavigate? : ((event: { event: Event, item: HTMLElement|null, oldItem: HTMLElement|null }) => void)|string
    /**
     * Triggered when a widget which had been in a non-visible state for any reason
     * achieves visibility.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/view/Calendar#event-paint)
     * @param {object} event Event object
     * @param {Widget} event.source The widget being painted.
     * @param {boolean} event.firstPaint `true` if this is the first paint.
     */
    onPaint? : ((event: { source: Widget, firstPaint: boolean }) => void)|string
    /**
     * Fires on the owning Scheduler after a paste action is performed.
     * @param {object} event Event object
     * @param {Scheduler} event.source Owner scheduler
     * @param {EventModel[]} event.eventRecords The pasted event records
     * @param {AssignmentModel[]} event.assignmentRecords The pasted assignment records
     * @param {EventModel[]} event.pastedEventRecords (DEPRECATED) Use eventRecords instead
     * @param {EventModel[]} event.originalEventRecords The original copied event records
     * @param {AssignmentModel[]} event.originalAssignmentRecords The original copied assignment records
     * @param {Date} event.date date Pasted to this date
     * @param {ResourceModel} event.resourceRecord The target resource record
     * @param {boolean} event.isCut `true` if this is a cut action
     * @param {string} event.entityName 'event' to distinguish this event from other paste events
     */
    onPaste? : ((event: { source: Scheduler, eventRecords: EventModel[], assignmentRecords: AssignmentModel[], pastedEventRecords: EventModel[], originalEventRecords: EventModel[], originalAssignmentRecords: AssignmentModel[], date: Date, resourceRecord: ResourceModel, isCut: boolean, entityName: string }) => void)|string
    /**
     * Fires on the owning Scheduler if a paste action is not allowed. Depending on if the EventStore is using
     * [singleAssignment](https://bryntum.com/products/calendar/docs/api/Scheduler/data/EventStore#config-singleAssignment) or the configuration of
     * [copyPasteAction](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/EventCopyPaste#config-copyPasteAction), either the `eventRecords` or the `assignmentRecords` param will be populated
     * with record copies.
     * @param {object} event Event object
     * @param {Scheduler} event.source Owner scheduler
     * @param {EventModel[]} event.eventRecords The event records about to be pasted
     * @param {AssignmentModel[]} event.assignmentRecords The assignment records about to be pasted
     * @param {EventModel[]} event.originalEventRecords The event records that were copied
     * @param {AssignmentModel[]} event.originalAssignmentRecords The assignment records that were copied
     * @param {Date} event.date The paste date
     * @param {ResourceModel} event.resourceRecord The target resource record
     * @param {boolean} event.isCut `true` if this is a cut action
     * @param {string} event.entityName 'event' to distinguish this event from other `pasteNotAllowed` events
     * @param {'overlappingEvents'|'resourceReadOnly'} event.reason A string id to use for displaying an error message to the user.
     */
    onPasteNotAllowed? : ((event: { source: Scheduler, eventRecords: EventModel[], assignmentRecords: AssignmentModel[], originalEventRecords: EventModel[], originalAssignmentRecords: AssignmentModel[], date: Date, resourceRecord: ResourceModel, isCut: boolean, entityName: string, reason: 'overlappingEvents'|'resourceReadOnly' }) => void)|string
    /**
     * Fired when a Widget's read only state is toggled
     * @param {object} event Event object
     * @param {boolean} event.readOnly Read only or not
     */
    onReadOnly? : ((event: { readOnly: boolean }) => void)|string
    /**
     * This event is fired after a widget's elements have been synchronized due to a direct or indirect call
     * to [recompose](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#function-recompose), if this results in some change to the widget's rendered DOM elements.
     */
    onRecompose? : (() => void)|string
    /**
     * Fires when one of the views in this Calendar refreshes.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which refreshed.
     */
    onRefresh? : ((event: { source: typeof CalendarMixin }) => void)|string
    /**
     * Fired when the encapsulating element of a Widget resizes *only when [monitorResize](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-monitorResize) is `true`*.
     * @param {object} event Event object
     * @param {Widget} event.source This Widget
     * @param {number} event.width The new width
     * @param {number} event.height The new height
     * @param {number} event.oldWidth The old width
     * @param {number} event.oldHeight The old height
     */
    onResize? : ((event: { source: Widget, width: number, height: number, oldWidth: number, oldHeight: number }) => void)|string
    /**
     * Triggered when a new [responsiveState](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Responsive#config-responsiveState) is applied.
     * @param {object} event Event object
     * @param {Widget} event.source The widget whose `responsiveState` has changed
     * @param {string} event.state The new value for the widget's `responsiveState`
     * @param {string} event.oldState The previous value for the widget's `responsiveState`
     */
    onResponsiveStateChange? : ((event: { source: Widget, state: string, oldState: string }) => void)|string
    /**
     * Fires when a click made its way through to an empty part of any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date clicked on. _Note that in a [DayView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView), this will include granular time information. For other views, this is the *start* of the date.
     */
    onScheduleClick? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date }) => void)|string
    /**
     * Fires when a context menu made its way through to an empty part of any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date clicked on. _Note that in a [DayView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView), this will include granular time information. For other views, this is the *start* of the date.
     */
    onScheduleContextMenu? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date }) => void)|string
    /**
     * Fires when a dblclick made its way through to an empty part of any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date double clicked on. _Note that in a [DayView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView), this will include granular time information. For other views, this is the *start* of the date.
     */
    onScheduleDblClick? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date }) => void)|string
    /**
     * This event fires on the owning Scheduler or Gantt widget before the context menu is shown for the schedule.
     * Allows manipulation of the items to show in the same way as in `processItems`. Returning `false` from a listener
     * prevents the menu from being shown.
     * @param {object} event Event object
     * @param {Scheduler} event.source
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {EventModel} event.eventRecord Event record for which the menu was triggered
     * @param {ResourceModel} event.resourceRecord Resource record
     * @param {Date} event.date Clicked date, rounded according to viewPreset's settings
     * @param {AssignmentModel} event.assignmentRecord Assignment record, if assignments are used
     * @param {HTMLElement} event.eventElement
     */
    onScheduleMenuBeforeShow? : ((event: { source: Scheduler, items: Record<string, MenuItemConfig>, eventRecord: EventModel, resourceRecord: ResourceModel, date: Date, assignmentRecord: AssignmentModel, eventElement: HTMLElement }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Scheduler or Gantt widget when an item is selected in the context menu.
     * @param {object} event Event object
     * @param {Scheduler} event.source
     * @param {MenuItem} event.item
     * @param {ResourceModel} event.resourceRecord
     * @param {Date} event.date Clicked date, rounded according to viewPreset's settings
     * @param {HTMLElement} event.element
     */
    onScheduleMenuItem? : ((event: { source: Scheduler, item: MenuItem, resourceRecord: ResourceModel, date: Date, element: HTMLElement }) => void)|string
    /**
     * This event fires on the owning Scheduler or Gantt widget after showing the context menu for the schedule.
     * @param {object} event Event object
     * @param {Scheduler} event.source
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {Menu} event.menu The menu
     * @param {ResourceModel} event.resourceRecord
     * @param {Date} event.date Clicked date, rounded according to viewPreset's settings
     * @param {HTMLElement} event.targetElement
     */
    onScheduleMenuShow? : ((event: { source: Scheduler, items: Record<string, MenuItemConfig>, menu: Menu, resourceRecord: ResourceModel, date: Date, targetElement: HTMLElement }) => void)|string
    /**
     * Fires when a mousedown made its way through to an empty part of any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date mousedowned on. _Note that in a [DayView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView), this will include granular time information. For other views, this is the *start* of the date.
     */
    onScheduleMouseDown? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date }) => void)|string
    /**
     * Fires when a mouseout made it's way through to an empty part of any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date mouseouted. _Note that in a [DayView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView), this will include granular time information. For other views, this is the *start* of the date.
     */
    onScheduleMouseOut? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date }) => void)|string
    /**
     * Fires when a mouseover made its way through to an empty part of any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date mouseovered. _Note that in a [DayView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView), this will include granular time information. For other views, this is the *start* of the date.
     */
    onScheduleMouseOver? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date }) => void)|string
    /**
     * Fires when a mouseup made its way through to an empty part of any view.
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {Date} event.date The date mouseupped on. _Note that in a [DayView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/DayView), this will include granular time information. For other views, this is the *start* of the date.
     */
    onScheduleMouseUp? : ((event: { source: typeof CalendarMixin, domEvent: Event, date: Date }) => void)|string
    /**
     * Fired any time there is a change to the events selected in the Calendar.
     * @param {object} event Event object
     * @param {Calendar} event.source The Calendar instance.
     * @param {'select'|'deselect'|'update'|'clear'} event.action One of the actions 'select', 'deselect', 'update', 'clear'
     * @param {EventModel[]} event.selected An array of the Events added to the selection.
     * @param {EventModel[]} event.deselected An array of the Event removed from the selection.
     * @param {EventModel[]} event.selection The new selection.
     */
    onSelectionChange? : ((event: { source: Calendar, action: 'select'|'deselect'|'update'|'clear', selected: EventModel[], deselected: EventModel[], selection: EventModel[] }) => void)|string
    /**
     * Triggered after a widget is shown.
     * @param {object} event Event object
     * @param {Widget} event.source The widget
     */
    onShow? : ((event: { source: Widget }) => void)|string
    /**
     * Fired when the sidebar is collapsed either through the UI or programatically.
     * @param {object} event Event object
     * @param {Calendar} event.source This Calendar instance
     */
    onSidebarCollapse? : ((event: { source: Calendar }) => void)|string
    /**
     * Fired when the sidebar is expanded either through the UI or programatically.
     * @param {object} event Event object
     * @param {Calendar} event.source This Calendar instance
     */
    onSidebarExpand? : ((event: { source: Calendar }) => void)|string
    /**
     * Fired on the owning Scheduler or Gantt widget when a click happens on a time range header element
     * @param {object} event Event object
     * @param {Scheduler} event.source Scheduler instance
     * @param {TimeSpan} event.timeRangeRecord The record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onTimeRangeHeaderClick? : ((event: { source: Scheduler, timeRangeRecord: TimeSpan, domEvent: MouseEvent }) => void)|string
    /**
     * Fired on the owning Scheduler or Gantt widget when a right click happens on a time range header element
     * @param {object} event Event object
     * @param {Scheduler} event.source Scheduler instance
     * @param {TimeSpan} event.timeRangeRecord The record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onTimeRangeHeaderContextMenu? : ((event: { source: Scheduler, timeRangeRecord: TimeSpan, domEvent: MouseEvent }) => void)|string
    /**
     * Fired on the owning Scheduler or Gantt widget when a double click happens on a time range header element
     * @param {object} event Event object
     * @param {Scheduler} event.source Scheduler instance
     * @param {TimeSpan} event.timeRangeRecord The record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onTimeRangeHeaderDblClick? : ((event: { source: Scheduler, timeRangeRecord: TimeSpan, domEvent: MouseEvent }) => void)|string
    /**
     * A header [tool](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-tools) has been clicked.
     * @param {object} event Event object
     * @param {Tool} event.source This Panel.
     * @param {Tool} event.tool The tool which is being clicked.
     */
    onToolClick? : ((event: { source: Tool, tool: Tool }) => void)|string
    /**
     * Fires when one of the child views is painted. That is when it becomes visible.
     * Note that due to the slide-in animation, while the view is visible, it will not
     * yet be in its final position.
     * @param {object} event Event object
     * @param {Widget} event.source The widget being painted
     * @param {boolean} event.firstPaint `true` if this is the first paint
     */
    onViewPaint? : ((event: { source: Widget, firstPaint: boolean }) => void)|string
    /**
     * Fires when a week number is clicked. This will include the week number cells of a
     * [YearView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/YearView), the week number in the first cell
     * in each row of a [MonthView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/MonthView)
     * @param {object} event Event object
     * @param {typeof CalendarMixin} event.source The view which triggered the event.
     * @param {Event} event.domEvent The initiating DOM event.
     * @param {number[]} event.week The `[year, week]` clicked on.
     * @param {Date} event.date The date clicked on.
     */
    onWeekNumberClick? : ((event: { source: typeof CalendarMixin, domEvent: Event, week: number[], date: Date }) => void)|string

    // Features
    /**
     * This feature provides drag-based event creation and modification for Calendars. When enabled (which is the default
     * for calendars), the user can do the following via the mouse or touch screen:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/CalendarDrag)
     */
    dragFeature? : object|boolean|string|CalendarDrag|CalendarDragConfig
    /**
     * A Feature that allows showing additional time before &amp; after an event, to visualize things like travel time - or the time you
     * need to prepare a room for a meeting + clean it up after.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/EventBuffer)
     */
    eventBufferFeature? : object|boolean|string|EventBuffer|EventBufferConfig
    /**
     * Allow using [Ctrl/CMD + C/X] and [Ctrl/CMD + V] to copy/cut and paste events.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/EventCopyPaste)
     */
    eventCopyPasteFeature? : object|boolean|string|EventCopyPaste|EventCopyPasteConfig
    /**
     * Feature that displays a popup containing fields for editing event data.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/EventEdit)
     */
    eventEditFeature? : object|boolean|string|EventEdit|EventEditConfig
    /**
     * A feature which shows a context menu when right-clicking events in the calendar.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/EventMenu)
     */
    eventMenuFeature? : object|boolean|string|EventMenu|EventMenuConfig
    /**
     * A feature which displays a [tooltip](https://bryntum.com/products/calendar/docs/api/Calendar/feature/EventTooltip#config-tooltip) containing extra information. The tooltip can
     * be triggered by clicking or hovering an event bar element (see [showOn](https://bryntum.com/products/calendar/docs/api/Calendar/feature/EventTooltip#config-showOn)).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/EventTooltip)
     */
    eventTooltipFeature? : object|boolean|string|EventTooltip|EventTooltipConfig
    /**
     * <strong>NOTE</strong>: This class requires a 3rd party library to operate.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/experimental/ExcelExporter)
     */
    excelExporterFeature? : object|boolean|string|ExcelExporter|ExcelExporterConfig
    /**
     * A Calendar feature which allows new events to be dragged into the Calendar from an external source.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ExternalEventSource)
     */
    externalEventSourceFeature? : object|boolean|string|ExternalEventSource|ExternalEventSourceConfig
    /**
     * Loads the host Calendar's [CrudManager](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/SchedulerStores#config-crudManager) on demand
     * as the date range required to produce the UI changes.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/LoadOnDemand)
     */
    loadOnDemandFeature? : object|boolean|string|LoadOnDemand|LoadOnDemandConfig
    /**
     * This feature enables printing of the current Calendar mode.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/print/Print)
     */
    printFeature? : object|boolean|string|Print|PrintConfig
    /**
     * A feature which adds a context menu to the unoccupied parts of calendar views.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ScheduleMenu)
     */
    scheduleMenuFeature? : object|boolean|string|ScheduleMenu|ScheduleMenuConfig
    /**
     * A feature that displays a tooltip containing the time at the mouse position when hovering empty parts of the schedule:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/ScheduleTooltip)
     */
    scheduleTooltipFeature? : object|boolean|string|ScheduleTooltip|ScheduleTooltipConfig
    /**
     * This feature provides an easy way to highlight ranges of time in a calendar's day and week views. Each time range is
     * represented using the [TimeRangeModel](https://bryntum.com/products/calendar/docs/api/Calendar/model/TimeRangeModel).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/TimeRanges)
     */
    timeRangesFeature? : object|boolean|string|TimeRanges|TimeRangesConfig
    /**
     * A Feature which, when present, adds a week row expansion tool to week rows in
     * [MonthView](https://bryntum.com/products/calendar/docs/api/Calendar/widget/MonthView)s so that a row which contains overflowing cells
     * can be expanded to show all events with no overflow.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Calendar/feature/WeekExpander)
     */
    weekExpanderFeature? : object|boolean|string|WeekExpander|WeekExpanderConfig

}

export class BryntumCalendar extends React.Component<BryntumCalendarProps> {

    static instanceClass = Calendar;

    static instanceName = 'Calendar';

    processWidgetContent = processWidgetContent;

    // Component is View class
    static isView = true;

    // Hidden div to host unused portals DOM
    portalsCache? : HTMLElement;

    // CSS class used for individual portal containers
    portalContainerClass = 'b-react-portal-container';

    state = {
        // Holds React portals
        portals : new Map(),

        // Needed to trigger refresh when portals change
        generation : 0
    };

    static featureNames = [
        'dragFeature',
        'eventBufferFeature',
        'eventCopyPasteFeature',
        'eventEditFeature',
        'eventMenuFeature',
        'eventTooltipFeature',
        'excelExporterFeature',
        'externalEventSourceFeature',
        'loadOnDemandFeature',
        'printFeature',
        'scheduleMenuFeature',
        'scheduleTooltipFeature',
        'timeRangesFeature',
        'weekExpanderFeature'
    ];

    projectStores = true;

    dataStores = {
        'assignmentStore': 'assignments',
        'dependencyStore': 'dependencies',
        'eventStore': 'events',
        'resourceStore': 'resources',
        'resourceTimeRangeStore': 'resourceTimeRanges',
        'timeRangeStore': 'timeRanges'
    };

    static configNames = [
        'adopt',
        'align',
        'anchor',
        'ariaDescription',
        'ariaLabel',
        'assignmentStore',
        'autoCreate',
        'bbar',
        'bodyCls',
        'bubbleEvents',
        'callOnFunctions',
        'centered',
        'collapsible',
        'config',
        'constrainTo',
        'contentElementCls',
        'contextMenuTriggerEvent',
        'crudManagerClass',
        'datePicker',
        'defaultBindProperty',
        'defaultFocus',
        'defaults',
        'deselectOnClick',
        'destroyStores',
        'dock',
        'draggable',
        'enableDeleteKey',
        'enableRecurringEvents',
        'enableUndoRedoKeys',
        'eventSelectionDisabled',
        'eventStore',
        'filterEventResources',
        'floating',
        'footer',
        'header',
        'hideAnimation',
        'hideWhenEmpty',
        'highlightDate',
        'highlightPredecessors',
        'highlightSuccessors',
        'ignoreParentReadOnly',
        'includeWeekendsButton',
        'itemCls',
        'lazyItems',
        'listeners',
        'loadMask',
        'loadMaskDefaults',
        'loadMaskError',
        'localeClass',
        'localizable',
        'localizableProperties',
        'maintainSelectionOnDatasetChange',
        'maskDefaults',
        'masked',
        'monitorResize',
        'multiEventSelect',
        'namedItems',
        'navigatorPlacement',
        'overlaySidebar',
        'owner',
        'positioned',
        'preventTooltipOnTouch',
        'relayStoreEvents',
        'resourceImagePath',
        'resourceStore',
        'resourceTimeRangeStore',
        'responsive',
        'responsiveRoot',
        'responsiveState',
        'responsiveTarget',
        'ripple',
        'rootElement',
        'scrollAction',
        'showAnimation',
        'showRecurringUI',
        'showTooltipWhenDisabled',
        'sidebar',
        'stateful',
        'statefulEvents',
        'stateId',
        'stateProvider',
        'strips',
        'syncMask',
        'tab',
        'tag',
        'tbar',
        'textAlign',
        'timeRangeStore',
        'timeZone',
        'trapFocus',
        'triggerSelectionChangeOnRemove',
        'type',
        'ui',
        'viewContainer',
        'weekStartDay',
        'weight'
    ];

    static propertyConfigNames = [
        'activeDateCls',
        'alignSelf',
        'appendTo',
        'ariaLive',
        'assignments',
        'cls',
        'collapsed',
        'crudManager',
        'dataset',
        'date',
        'dateFormat',
        'defaultCalendar',
        'deselectAllOnScheduleClick',
        'disabled',
        'events',
        'extraData',
        'flex',
        'height',
        'hidden',
        'hideNonWorkingDays',
        'id',
        'insertBefore',
        'insertFirst',
        'isEventSelectable',
        'items',
        'keyMap',
        'layout',
        'layoutStyle',
        'margin',
        'maxDate',
        'maxHeight',
        'maximizeOnMobile',
        'maxWidth',
        'minDate',
        'minHeight',
        'minWidth',
        'mode',
        'modeDefaults',
        'modes',
        'nonWorkingDays',
        'onActiveItemChange',
        'onAfterEventEdit',
        'onAfterEventSave',
        'onBeforeActiveItemChange',
        'onBeforeAssignmentDelete',
        'onBeforeAssignmentSelectionChange',
        'onBeforeAutoCreate',
        'onBeforeCopy',
        'onBeforeCreateMode',
        'onBeforeCSVExport',
        'onBeforeDestroy',
        'onBeforeDragCreate',
        'onBeforeDragCreateEnd',
        'onBeforeDragMove',
        'onBeforeDragMoveEnd',
        'onBeforeDragResize',
        'onBeforeDragResizeEnd',
        'onBeforeDropExternal',
        'onBeforeEventDelete',
        'onBeforeEventEdit',
        'onBeforeEventEditShow',
        'onBeforeEventSave',
        'onBeforeEventSelectionChange',
        'onBeforeExcelExport',
        'onBeforeHide',
        'onBeforePaste',
        'onBeforePrint',
        'onBeforeResponsiveStateChange',
        'onBeforeSetRecord',
        'onBeforeShow',
        'onBeforeStateApply',
        'onBeforeStateSave',
        'onCatchAll',
        'onCellOverflowClick',
        'onCollapse',
        'onContextMenuItem',
        'onContextMenuToggleItem',
        'onCopy',
        'onCurrentTimelineUpdate',
        'onDataChange',
        'onDateChange',
        'onDateRangeChange',
        'onDateRangeLoad',
        'onDateRangeRequested',
        'onDayNumberClick',
        'onDestroy',
        'onDirtyStateChange',
        'onDragCreateEnd',
        'onDragMoveEnd',
        'onDragMoveExternal',
        'onDragResizeEnd',
        'onDropExternal',
        'onEmptyCellClick',
        'onEventAutoCreated',
        'onEventClick',
        'onEventContextMenu',
        'onEventDblClick',
        'onEventEditBeforeSetRecord',
        'onEventKeyDown',
        'onEventMenuBeforeShow',
        'onEventMenuItem',
        'onEventMenuShow',
        'onEventMouseDown',
        'onEventMouseEnter',
        'onEventMouseLeave',
        'onEventMouseOut',
        'onEventMouseOver',
        'onEventMouseUp',
        'onEventPropagate',
        'onExpand',
        'onFocusIn',
        'onFocusOut',
        'onHide',
        'onLoadOnDemandFail',
        'onMonthNameClick',
        'onNavigate',
        'onPaint',
        'onPaste',
        'onPasteNotAllowed',
        'onReadOnly',
        'onRecompose',
        'onRefresh',
        'onResize',
        'onResponsiveStateChange',
        'onScheduleClick',
        'onScheduleContextMenu',
        'onScheduleDblClick',
        'onScheduleMenuBeforeShow',
        'onScheduleMenuItem',
        'onScheduleMenuShow',
        'onScheduleMouseDown',
        'onScheduleMouseOut',
        'onScheduleMouseOver',
        'onScheduleMouseUp',
        'onSelectionChange',
        'onShow',
        'onSidebarCollapse',
        'onSidebarExpand',
        'onTimeRangeHeaderClick',
        'onTimeRangeHeaderContextMenu',
        'onTimeRangeHeaderDblClick',
        'onToolClick',
        'onViewPaint',
        'onWeekNumberClick',
        'project',
        'readOnly',
        'recurrenceConfirmationPopup',
        'resources',
        'resourceTimeRanges',
        'scrollable',
        'selectResourceOnEventNavigate',
        'selectResourceOnScheduleClick',
        'strictRecordMapping',
        'timeRanges',
        'title',
        'tools',
        'tooltip',
        'useContextualRecurrenceRules',
        'width',
        'x',
        'y'
    ];

    static propertyNames = [
        'accessibility',
        'activeSubView',
        'activeView',
        'anchorSize',
        'hasChanges',
        'isEngineReady',
        'isSettingValues',
        'isValid',
        'parent',
        'record',
        'selectedEvents',
        'state',
        'values',
        'views'
    ];

    // Component instance
    instance!: Calendar;

    // Component element
    element! : HTMLElement;

    componentDidMount(): void {
        this.instance = createWidget(this);
    }

    componentWillUnmount(): void {
        // @ts-ignore
        this.instance?.destroy?.();
    }

    /**
     * Component about to be updated, from changing a prop using state.
     * React to it depending on what changed and prevent react from re-rendering our component.
     * @param nextProps
     * @param nextState
     * @returns {boolean}
     */
    shouldComponentUpdate(nextProps: Readonly<BryntumCalendarProps>, nextState: Readonly<{}>): boolean {
        return shouldComponentUpdate(this, nextProps, nextState);
    }

    render(): React.ReactNode {

        const className = `b-react-calendar-container`;
        return (
            <React.Fragment>
                <div className={className} ref={(element) => (this.element = element!)}>
                    {Array.from(this.state.portals).map((item) => item[1])}
                </div>
                <div ref={el => (this.portalsCache = el!)} className="b-react-portals-cache" style={{ display : 'none' }}></div>
            </React.Fragment>
        );

    }
}
